{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Plan\n",
    "- Similar to 20160720_mendelian_error_duplicate_concordance but\n",
    "- Split multiallelics in biallelic+spanning deletion and \"true\" multiallelic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See 20160203_release5_npy_hdf5.ipynb for creation of VCF specific to crosses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python 3.4.3 |Anaconda 2.2.0 (64-bit)| (default, Mar  6 2015, 12:03:53) \n",
      "[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]\n",
      "numpy 1.9.2\n",
      "scipy 0.15.1\n",
      "pandas 0.15.2\n",
      "numexpr 2.3.1\n",
      "pysam 0.8.3\n",
      "pysamstats 0.23\n",
      "petl 1.0.11\n",
      "petlx 1.0.3\n",
      "vcf 0.6.7\n",
      "vcfnp 2.3.0.dev0\n",
      "h5py 2.4.0\n",
      "tables 3.1.1\n"
     ]
    }
   ],
   "source": [
    "%run _standard_imports.ipynb\n",
    "%run _shared_setup.ipynb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "release5_final_files_dir = '/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0'\n",
    "# chrom_vcf_fn = \"%s/SNP_INDEL_Pf3D7_14_v3.combined.filtered.vcf.gz\" % (release5_final_files_dir)\n",
    "crosses_vcf_fn = \"%s/SNP_INDEL_WG.combined.filtered.crosses.vcf.gz\" % (release5_final_files_dir)\n",
    "sites_only_vcf_fn = \"%s/SNP_INDEL_WG.combined.filtered.sites.vcf.gz\" % (release5_final_files_dir)\n",
    "\n",
    "output_dir = \"/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance\"\n",
    "!mkdir -p {output_dir}/vcf\n",
    "\n",
    "release5_crosses_metadata_txt_fn = '../../meta/pf3k_release_5_crosses_metadata.txt'\n",
    "gff_fn = \"%s/Pfalciparum.noseq.gff3.gz\" % output_dir\n",
    "cds_gff_fn = \"%s/Pfalciparum.noseq.gff3.cds.gz\" % output_dir\n",
    "\n",
    "results_table_fn = \"%s/genotype_quality.xlsx\" % output_dir\n",
    "counts_table_fn = \"%s/variant_counts.xlsx\" % output_dir\n",
    "\n",
    "BCFTOOLS = '/nfs/team112_internal/rp7/src/github/malariagen/methods-dev/pf3k_techbm/opt_4/bcftools/bcftools'\n",
    "GATK = '/software/jre1.7.0_25/bin/java -jar /nfs/team112_internal/production/tools/bin/gatk/GenomeAnalysisTK-3.4-46/GenomeAnalysisTK.jar'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/Pfalciparum.noseq.gff3.gz'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2016-07-21 17:47:34--  ftp://ftp.sanger.ac.uk/pub/project/pathogens/gff3/2016-06/Pfalciparum.noseq.gff3.gz\n",
      "           => `/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/Pfalciparum.noseq.gff3.gz'\n",
      "Resolving ftp.sanger.ac.uk (ftp.sanger.ac.uk)... 193.62.203.17\n",
      "Connecting to ftp.sanger.ac.uk (ftp.sanger.ac.uk)|193.62.203.17|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/project/pathogens/gff3/2016-06 ... done.\n",
      "==> SIZE Pfalciparum.noseq.gff3.gz ... 4852221\n",
      "==> PASV ... done.    ==> RETR Pfalciparum.noseq.gff3.gz ... done.\n",
      "Length: 4852221 (4.6M) (unauthoritative)\n",
      "\n",
      "100%[======================================>] 4,852,221   --.-K/s   in 0.09s   \n",
      "\n",
      "2016-07-21 17:47:34 (54.1 MB/s) - `/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/Pfalciparum.noseq.gff3.gz' saved [4852221]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.sanger.ac.uk/pub/project/pathogens/gff3/2016-06/Pfalciparum.noseq.gff3.gz \\\n",
    "    -O {gff_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!zgrep CDS {gff_fn} | sort -k1,1 -k4n,5n | cut -f 1,4,5 | sed 's/$/\\t1/' | bgzip -c > {cds_gff_fn} && tabix -s1 -b2 -e3 {cds_gff_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/SNP_INDEL_WG.combined.filtered.crosses.vcf.gz'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crosses_vcf_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "multiallelic_header_fn = \"%s/vcf/MULTIALLELIC.hdr\" % (output_dir)\n",
    "fo=open(multiallelic_header_fn, 'w')\n",
    "print('##INFO=<ID=MULTIALLELIC,Number=1,Type=String,Description=\"Is position biallelic (BI), biallelic plus spanning deletion (SD) or truly multiallelic (MU)\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "variant_type_header_fn = \"%s/vcf/VARIANT_TYPE.hdr\" % (output_dir)\n",
    "fo=open(variant_type_header_fn, 'w')\n",
    "print('##INFO=<ID=VARIANT_TYPE,Number=1,Type=String,Description=\"SNP or indel (IND)\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cds_header_fn = \"%s/vcf/CDS.hdr\" % (output_dir)\n",
    "fo=open(cds_header_fn, 'w')\n",
    "print('##INFO=<ID=CDS,Number=0,Type=Flag,Description=\"Is position coding\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_analysis_vcf(input_vcf_fn=crosses_vcf_fn, region='Pf3D7_14_v3:1000000-1100000',\n",
    "                        output_vcf_fn=None, BCFTOOLS=BCFTOOLS, rewrite=False):\n",
    "    if output_vcf_fn is None:\n",
    "        if region is None:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir)\n",
    "        else:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_%s.analysis.vcf.gz\" % (output_dir, region)\n",
    "    intermediate_fns = collections.OrderedDict()\n",
    "    for intermediate_file in ['nonref', 'multiallelic', 'triallelic', 'bi_allelic', 'spanning_deletion', 'triallelic_no_sd', 'multiallelics',\n",
    "                              'biallelic', 'str', 'snps', 'indels', 'strs', 'variant_type', 'coding', 'analysis',\n",
    "                             'site_snps', 'site_indels', 'site_variant_type', 'site_analysis']:\n",
    "        intermediate_fns[intermediate_file] = output_vcf_fn.replace('analysis', intermediate_file)\n",
    "    \n",
    "    if rewrite or not os.path.exists(intermediate_fns['nonref']):\n",
    "        if region is not None:\n",
    "            !{BCFTOOLS} annotate --regions {region} --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set,\\\n",
    "FORMAT/PGT,FORMAT/PID,FORMAT/PL {input_vcf_fn} | \\\n",
    "            {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['nonref']}\n",
    "        else:\n",
    "            !{BCFTOOLS} annotate --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set,\\\n",
    "FORMAT/PGT,FORMAT/PID,FORMAT/PL {input_vcf_fn} | \\\n",
    "            {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['nonref']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['nonref']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tMU\\n' --include 'N_ALT>2' {intermediate_fns['nonref']} | bgzip -c > {intermediate_fns['multiallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['multiallelic']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['triallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t%ALT\\tSD\\n' --include 'N_ALT=2' {intermediate_fns['nonref']} | bgzip -c > {intermediate_fns['triallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['triallelic']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['bi_allelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t%ALT\\tBI\\n' --include 'N_ALT=1' {intermediate_fns['nonref']} | bgzip -c > {intermediate_fns['bi_allelic']} && tabix -s1 -b2 -e2 {intermediate_fns['bi_allelic']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['spanning_deletion']):\n",
    "        !zgrep '\\*' {intermediate_fns['triallelic']} | bgzip -c > {intermediate_fns['spanning_deletion']} && tabix -s1 -b2 -e2 {intermediate_fns['spanning_deletion']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['triallelic_no_sd']):\n",
    "        !zgrep -v '\\*' {intermediate_fns['triallelic']} | sed 's/SD/MU/g' | bgzip -c > {intermediate_fns['triallelic_no_sd']} && tabix -s1 -b2 -e2 {intermediate_fns['triallelic_no_sd']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelics']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['bi_allelic']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "        -h {multiallelic_header_fn} {intermediate_fns['nonref']} | \\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['spanning_deletion']} -c CHROM,POS,-,INFO/MULTIALLELIC |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['multiallelic']} -c CHROM,POS,INFO/MULTIALLELIC |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['triallelic_no_sd']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "        -Oz -o {intermediate_fns['multiallelics']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['multiallelics']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['biallelic']):\n",
    "        !{BCFTOOLS} norm -m -any --fasta-ref {GENOME_FN} {intermediate_fns['multiallelics']} | \\\n",
    "        {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['biallelic']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['biallelic']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['str']):\n",
    "#         !{GATK} -T VariantAnnotator \\\n",
    "#             -R {GENOME_FN} \\\n",
    "#             -o {intermediate_fns['str']} \\\n",
    "#             -A TandemRepeatAnnotator  \\\n",
    "#             -V {intermediate_fns['biallelic']}\n",
    "# #         !{BCFTOOLS} index --tbi {intermediate_fns['str']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['snps']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSNP\\n' --include 'TYPE=\"snp\"' {intermediate_fns['biallelic']} | bgzip -c > {intermediate_fns['snps']} && tabix -s1 -b2 -e2 -f {intermediate_fns['snps']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['indels']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tINDEL\\n' --include 'TYPE!=\"snp\"' {intermediate_fns['biallelic']} | bgzip -c > {intermediate_fns['indels']} && tabix -s1 -b2 -e2 -f {intermediate_fns['indels']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['strs']):\n",
    "#         !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSTR\\n' --include 'TYPE!=\"snp\" && STR=1' {intermediate_fns['str']} | bgzip -c > {intermediate_fns['strs']} && tabix -s1 -b2 -e2 -f {intermediate_fns['strs']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['variant_type']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['snps']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -h {variant_type_header_fn} {intermediate_fns['biallelic']} |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['indels']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -Oz -o {intermediate_fns['variant_type']} \n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['variant_type']}\n",
    "\n",
    "        #         {BCFTOOLS} annotate -a {intermediate_fns['strs']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['analysis']):\n",
    "        !{BCFTOOLS} annotate -a {cds_gff_fn} -c CHROM,FROM,TO,CDS \\\n",
    "        -h {cds_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['analysis']} {intermediate_fns['variant_type']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['analysis']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['site_snps']):\n",
    "#         !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSNP\\n' --include 'TYPE=\"snp\"' {intermediate_fns['multiallelics']} | bgzip -c > {intermediate_fns['site_snps']} && tabix -s1 -b2 -e2 -f {intermediate_fns['site_snps']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['site_indels']):\n",
    "#         !{BCFTOOLS} query -f'%CHROM\\t%POS\\tINDEL\\n' --include 'TYPE!=\"snp\"' {intermediate_fns['multiallelics']} | bgzip -c > {intermediate_fns['site_indels']} && tabix -s1 -b2 -e2 -f {intermediate_fns['site_indels']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['site_variant_type']):\n",
    "#         !{BCFTOOLS} annotate -a {intermediate_fns['site_snps']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "#         -h {variant_type_header_fn} {intermediate_fns['multiallelics']} | \\\n",
    "#         {BCFTOOLS} annotate -a {intermediate_fns['site_indels']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "#         -Oz -o {intermediate_fns['site_variant_type']} \n",
    "#         !{BCFTOOLS} index --tbi {intermediate_fns['site_variant_type']}\n",
    "\n",
    "#     if rewrite or not os.path.exists(intermediate_fns['site_analysis']):\n",
    "#         !{BCFTOOLS} annotate -a {cds_gff_fn} -c CHROM,FROM,TO,CDS \\\n",
    "#         -h {cds_header_fn} \\\n",
    "#         -Oz -o {intermediate_fns['site_analysis']} {intermediate_fns['site_variant_type']}\n",
    "#         !{BCFTOOLS} index --tbi {intermediate_fns['site_analysis']}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# create_analysis_vcf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "create_analysis_vcf(region=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "98\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>sample</th>\n",
       "<th>acc</th>\n",
       "<th>study</th>\n",
       "<th>study_title</th>\n",
       "<th>clone</th>\n",
       "<th>parent_or_progeny</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>PG0008-CW</td>\n",
       "<td>ERS010540</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>DD2</td>\n",
       "<td>parent</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0004-CW</td>\n",
       "<td>ERS010539</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>HB3</td>\n",
       "<td>parent</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0023-C</td>\n",
       "<td>ERS010581,ERS010543</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>1BB5</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0022-Cx</td>\n",
       "<td>ERS088714</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>3BA6</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0024-C</td>\n",
       "<td>ERS010150</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>3BD5</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| sample      | acc                   | study  | study_title             | clone  | parent_or_progeny |\n",
       "+=============+=======================+========+=========================+========+===================+\n",
       "| 'PG0008-CW' | 'ERS010540'           | '1041' | 'HB3xDD2 cross progeny' | 'DD2'  | 'parent'          |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0004-CW' | 'ERS010539'           | '1041' | 'HB3xDD2 cross progeny' | 'HB3'  | 'parent'          |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0023-C'  | 'ERS010581,ERS010543' | '1041' | 'HB3xDD2 cross progeny' | '1BB5' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0022-Cx' | 'ERS088714'           | '1041' | 'HB3xDD2 cross progeny' | '3BA6' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0024-C'  | 'ERS010150'           | '1041' | 'HB3xDD2 cross progeny' | '3BD5' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "..."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_release5_crosses_metadata = etl.fromtsv(release5_crosses_metadata_txt_fn)\n",
    "print(len(tbl_release5_crosses_metadata.data()))\n",
    "tbl_release5_crosses_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "replicates_first = [\n",
    " 'PG0112-C',\n",
    " 'PG0062-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0055-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0004-CW',\n",
    " 'PG0111-C',\n",
    " 'PG0100-C',\n",
    " 'PG0079-C',\n",
    " 'PG0104-C',\n",
    " 'PG0086-C',\n",
    " 'PG0095-C',\n",
    " 'PG0078-C',\n",
    " 'PG0105-C',\n",
    " 'PG0102-C']\n",
    "\n",
    "replicates_second = [\n",
    " 'PG0112-CW',\n",
    " 'PG0065-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0067-C',\n",
    " 'PG0052-C',\n",
    " 'PG0111-CW',\n",
    " 'PG0100-CW',\n",
    " 'PG0079-CW',\n",
    " 'PG0104-CW',\n",
    " 'PG0086-CW',\n",
    " 'PG0095-CW',\n",
    " 'PG0078-CW',\n",
    " 'PG0105-CW',\n",
    " 'PG0102-CW']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "quads_first = [\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0055-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C']\n",
    "\n",
    "quads_second = [\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0067-C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Note the version created in this notebook doesn't work. I think this is because of R in Number of FORMAT field for AD,\n",
    "# which is part of spec for v4.2, but think GATK must have got rid of this in previous notebook\n",
    "analysis_vcf_fn = \"/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.analysis.vcf.gz\"\n",
    "vcf_reader = vcf.Reader(filename=analysis_vcf_fn)\n",
    "sample_ids = np.array(vcf_reader.samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "[96 48 40 40 40 42 42 43  2 94 82 63 87 69 76 61 89 84]\n",
      "18\n",
      "[97 51 42 43 53 43 53 53 39 95 83 64 88 70 77 62 90 85]\n"
     ]
    }
   ],
   "source": [
    "# sample_ids = tbl_release5_crosses_metadata.values('sample').array()\n",
    "\n",
    "rep_index_first = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], replicates_first)]\n",
    "rep_index_second = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], replicates_second)]\n",
    "print(len(rep_index_first))\n",
    "print(rep_index_first)\n",
    "print(len(rep_index_second))\n",
    "print(rep_index_second)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n",
      "[40 40 40 42 42 43]\n",
      "6\n",
      "[42 43 53 43 53 53]\n"
     ]
    }
   ],
   "source": [
    "# sample_ids = tbl_release5_crosses_metadata.values('sample').array()\n",
    "\n",
    "quad_index_first = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], quads_first)]\n",
    "quad_index_second = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], quads_second)]\n",
    "print(len(quad_index_first))\n",
    "print(quad_index_first)\n",
    "print(len(quad_index_second))\n",
    "print(quad_index_second)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3D7xHB3 cross progeny', '7G8xGB4 cross progeny',\n",
       "       'HB3xDD2 cross progeny'], \n",
       "      dtype='<U21')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_release5_crosses_metadata.distinct('study_title').values('study_title').array()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_variants_npy(vcf_fn):\n",
    "    output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    vcfnp.variants(\n",
    "        vcf_fn,\n",
    "#         fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE', 'RU',\n",
    "#                 'SNPEFF_EFFECT', 'AC', 'AN', 'RPA', 'CDS', 'MULTIALLELIC',\n",
    "#                 'VQSLOD', 'FILTER'],\n",
    "        fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE',\n",
    "                'SNPEFF_EFFECT', 'AC', 'AN', 'CDS', 'MULTIALLELIC',\n",
    "                'VQSLOD', 'FILTER'],\n",
    "        dtypes={\n",
    "            'REF':                      'a10',\n",
    "            'ALT':                      'a10',\n",
    "            'RegionType':               'a25',\n",
    "            'VariantType':              'a40',\n",
    "            'VARIANT_TYPE':             'a3',\n",
    "            'RU':                       'a40',\n",
    "            'SNPEFF_EFFECT':            'a33',\n",
    "            'CDS':                      bool,\n",
    "            'MULTIALLELIC':             'a2',\n",
    "        },\n",
    "        arities={\n",
    "            'ALT':   1,\n",
    "            'AF':    1,\n",
    "            'AC':    1,\n",
    "            'RPA':   2,\n",
    "            'ANN':   1,\n",
    "        },\n",
    "        fills={\n",
    "            'VQSLOD': np.nan,\n",
    "        },\n",
    "        flatten_filter=True,\n",
    "        progress=100000,\n",
    "        verbose=True,\n",
    "        cache=True,\n",
    "        cachedir=output_dir\n",
    "    )\n",
    "\n",
    "def create_calldata_npy(vcf_fn, max_alleles=2):\n",
    "    output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    vcfnp.calldata_2d(\n",
    "        vcf_fn,\n",
    "        fields=['GT', 'AD', 'DP', 'GQ'],\n",
    "        dtypes={\n",
    "            'AD': 'u2',\n",
    "        },\n",
    "        arities={\n",
    "            'AD': max_alleles,\n",
    "        },\n",
    "        progress=100000,\n",
    "        verbose=True,\n",
    "        cache=True,\n",
    "        cachedir=output_dir\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[vcfnp] 2016-07-21 18:02:51.213098 :: caching is enabled\n",
      "[vcfnp] 2016-07-21 18:02:51.214485 :: cache file available\n",
      "[vcfnp] 2016-07-21 18:02:51.215060 :: loading from cache file /lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.analysis.vcf.gz.vcfnp_cache/variants.npy\n",
      "[vcfnp] 2016-07-21 18:02:51.365817 :: caching is enabled\n",
      "[vcfnp] 2016-07-21 18:02:51.367831 :: cache file available\n",
      "[vcfnp] 2016-07-21 18:02:51.369112 :: loading from cache file /lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.analysis.vcf.gz.vcfnp_cache/calldata_2d.npy\n"
     ]
    }
   ],
   "source": [
    "# create_analysis_vcf(region='Pf3D7_14_v3', rewrite=True)\n",
    "# create_variants_npy(\"%s/vcf/SNP_INDEL_Pf3D7_14_v3:1000000-1100000.coding.vcf.gz\" % (output_dir))\n",
    "# create_calldata_npy(\"%s/vcf/SNP_INDEL_Pf3D7_14_v3:1000000-1100000.coding.vcf.gz\" % (output_dir))\n",
    "create_variants_npy(\"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir))\n",
    "create_calldata_npy(\"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# def create_variants_npy_2(vcf_fn):\n",
    "#     output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "#     if not os.path.exists(output_dir):\n",
    "#         os.makedirs(output_dir)\n",
    "#     vcfnp.variants(\n",
    "#         vcf_fn,\n",
    "#         fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE',\n",
    "#                 'SNPEFF_EFFECT', 'AC', 'AN', 'CDS', 'MULTIALLELIC',\n",
    "#                 'VQSLOD', 'FILTER'],\n",
    "#         dtypes={\n",
    "#             'REF':                      'a10',\n",
    "#             'ALT':                      'a10',\n",
    "#             'RegionType':               'a25',\n",
    "#             'VariantType':              'a40',\n",
    "#             'VARIANT_TYPE':             'a3',\n",
    "#             'SNPEFF_EFFECT':            'a33',\n",
    "#             'CDS':                      bool,\n",
    "#             'MULTIALLELIC':             'a2',\n",
    "#         },\n",
    "#         arities={\n",
    "#             'ALT':   1,\n",
    "#             'AF':    1,\n",
    "#             'AC':    1,\n",
    "#             'ANN':   1,\n",
    "#         },\n",
    "#         fills={\n",
    "#             'VQSLOD': np.nan,\n",
    "#         },\n",
    "#         flatten_filter=True,\n",
    "#         progress=100000,\n",
    "#         verbose=True,\n",
    "#         cache=True,\n",
    "#         cachedir=output_dir\n",
    "#     )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# create_variants_npy_2(\"%s/vcf/SNP_INDEL_WG.site_analysis.vcf.gz\" % (output_dir))\n",
    "# create_calldata_npy(\"%s/vcf/SNP_INDEL_WG.site_analysis.vcf.gz\" % (output_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "analysis_vcf_fn = \"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir)\n",
    "variants = np.load(\"%s.vcfnp_cache/variants.npy\" % analysis_vcf_fn)\n",
    "calldata = np.load(\"%s.vcfnp_cache/calldata_2d.npy\" % analysis_vcf_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([b'IND', b'SNP'], \n",
      "      dtype='|S3'), array([417635, 415108]))\n"
     ]
    }
   ],
   "source": [
    "print(np.unique(variants['VARIANT_TYPE'], return_counts=True))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# site_analysis_vcf_fn = \"%s/vcf/SNP_INDEL_WG.site_analysis.vcf.gz\" % (output_dir)\n",
    "# site_variants = np.load(\"%s.vcfnp_cache/variants.npy\" % site_analysis_vcf_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# print(np.unique(site_variants['VARIANT_TYPE'], return_counts=True))\n",
    "# print(np.unique(site_variants['MULTIALLELIC'], return_counts=True))\n",
    "# print(np.unique(site_variants['CDS'], return_counts=True))\n",
    "# print(np.unique(site_variants['FILTER_PASS'], return_counts=True))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([b'BI', b'MU', b'SD'], \n",
       "       dtype='|S2'), array([ 86374, 659316,  87053]))"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(variants['MULTIALLELIC'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def genotype_concordance_gatk(calldata=calldata,\n",
    "                              ix = ((variants['VARIANT_TYPE'] == b'SNP') & (variants['MULTIALLELIC'] == b'BI') &\n",
    "                                    (variants['CDS']) & variants['FILTER_PASS']),\n",
    "                              GQ_threshold=30,\n",
    "                              rep_index_first=rep_index_first, rep_index_second=rep_index_second,\n",
    "                              verbose=False):\n",
    "    GT = calldata['GT'][ix, :]\n",
    "    GT[calldata['GQ'][ix, :] < GQ_threshold] = b'./.'\n",
    "    \n",
    "    all_samples = sample_ids\n",
    "#     all_samples = tbl_release5_crosses_metadata.values('sample').array()\n",
    "    total_mendelian_errors = 0\n",
    "    total_homozygotes = 0\n",
    "    for cross in tbl_release5_crosses_metadata.distinct('study_title').values('study_title').array():\n",
    "        parents = (tbl_release5_crosses_metadata\n",
    "            .selecteq('study_title', cross)\n",
    "            .selecteq('parent_or_progeny', 'parent')\n",
    "            .values('sample')\n",
    "            .array()\n",
    "        )\n",
    "        parent1_calls = GT[:, all_samples==parents[0]]\n",
    "#         print(parent1_calls.shape)\n",
    "#         print(parent1_calls == b'1')\n",
    "        parent2_calls = GT[:, all_samples==parents[1]]\n",
    "#         print(parent1_calls.shape)\n",
    "        progeny = (tbl_release5_crosses_metadata\n",
    "            .selecteq('study_title', cross)\n",
    "            .selecteq('parent_or_progeny', 'progeny')\n",
    "            .values('sample')\n",
    "            .array()\n",
    "        )\n",
    "        mendelian_errors = np.zeros(len(progeny))\n",
    "        homozygotes = np.zeros(len(progeny))\n",
    "        for i, ox_code in enumerate(progeny):\n",
    "            progeny_calls = GT[:, all_samples==ox_code]\n",
    "#             print(ox_code, progeny_calls.shape)\n",
    "            error_calls = (\n",
    "                ((parent1_calls == b'0/0') & (parent2_calls == b'0/0') & (progeny_calls == b'1/1')) |\n",
    "                ((parent1_calls == b'1/1') & (parent2_calls == b'1/1') & (progeny_calls == b'0/0'))\n",
    "            )\n",
    "            homozygous_calls = (\n",
    "                ((parent1_calls == b'0/0') | (parent1_calls == b'1/1' )) &\n",
    "                ((parent2_calls == b'0/0') | (parent2_calls == b'1/1' )) &\n",
    "                ((progeny_calls == b'0/0') | (progeny_calls == b'1/1' ))\n",
    "            )\n",
    "            mendelian_errors[i] = np.sum(error_calls)\n",
    "            homozygotes[i] = np.sum(homozygous_calls)\n",
    "#         print(cross, mendelian_errors, homozygotes)\n",
    "        total_mendelian_errors = total_mendelian_errors + np.sum(mendelian_errors)\n",
    "        total_homozygotes = total_homozygotes + np.sum(homozygotes)\n",
    "        \n",
    "    mendelian_error_rate = total_mendelian_errors / total_homozygotes\n",
    "    \n",
    "    GT_both = (np.in1d(GT[:, rep_index_first], [b'0/0', b'1/1']) &\n",
    "                     np.in1d(GT[:, rep_index_second], [b'0/0', b'1/1'])\n",
    "                    )\n",
    "    GT_both = (\n",
    "        ((GT[:, rep_index_first] == b'0/0') | (GT[:, rep_index_first] == b'1/1')) &\n",
    "        ((GT[:, rep_index_second] == b'0/0') | (GT[:, rep_index_second] == b'1/1'))\n",
    "    )\n",
    "    GT_discordant = (\n",
    "        ((GT[:, rep_index_first] == b'0/0') & (GT[:, rep_index_second] == b'1/1')) |\n",
    "        ((GT[:, rep_index_first] == b'1/1') & (GT[:, rep_index_second] == b'0/0'))\n",
    "    )\n",
    "    missingness_per_sample = np.sum(GT == b'./.', 0)\n",
    "    if verbose:\n",
    "        print(missingness_per_sample)\n",
    "    mean_missingness = np.sum(missingness_per_sample) / (GT.shape[0] * GT.shape[1])\n",
    "    heterozygosity_per_sample = np.sum(GT == b'0/1', 0)\n",
    "    if verbose:\n",
    "        print(heterozygosity_per_sample)\n",
    "#     print(heterozygosity_per_sample)\n",
    "    mean_heterozygosity = np.sum(heterozygosity_per_sample) / (GT.shape[0] * GT.shape[1])\n",
    "#     print(calldata_both.shape)\n",
    "#     print(calldata_discordant.shape)\n",
    "    num_discordance_per_sample_pair = np.sum(GT_discordant, 0)\n",
    "    num_both_calls_per_sample_pair = np.sum(GT_both, 0)\n",
    "    if verbose:\n",
    "        print(num_discordance_per_sample_pair)\n",
    "        print(num_both_calls_per_sample_pair)\n",
    "    prop_discordances_per_sample_pair = num_discordance_per_sample_pair/num_both_calls_per_sample_pair\n",
    "    num_discordance_per_snp = np.sum(GT_discordant, 1)\n",
    "#     print(num_discordance_per_snp)\n",
    "#     print(variants_SNP_BIALLELIC[num_discordance_per_snp > 0])\n",
    "    mean_prop_discordances = (np.sum(num_discordance_per_sample_pair) / np.sum(num_both_calls_per_sample_pair))\n",
    "    num_of_alleles = np.sum(ix)\n",
    "    return(\n",
    "#         num_of_alleles,\n",
    "        mean_missingness,\n",
    "        mean_heterozygosity,\n",
    "        mean_prop_discordances,\n",
    "        mendelian_error_rate,\n",
    "#         prop_discordances_per_sample_pair,\n",
    "#         GT.shape\n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.052356625919844309,\n",
       " 0.0032449761513800922,\n",
       " 1.4812620352540365e-05,\n",
       " 3.1193964068578418e-05)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.017184907168486807,\n",
       " 0.0042406234524461125,\n",
       " 1.4499548097417631e-05,\n",
       " 3.1662086857739367e-05)"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.017230241313679999,\n",
       " 0.0042580681013045111,\n",
       " 1.449226354664335e-05,\n",
       " 3.1646874583436029e-05)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.017184907168486807,\n",
       " 0.0042406234524461125,\n",
       " 1.4499548097417631e-05,\n",
       " 3.1662086857739367e-05)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 True b'SNP' True b'BI'\n",
      "0 True b'SNP' True b'SD'\n",
      "0 True b'SNP' True b'MU'\n",
      "0 True b'SNP' False b'BI'\n",
      "0 True b'SNP' False b'SD'\n",
      "0 True b'SNP' False b'MU'\n",
      "0 True b'IND' True b'BI'\n",
      "0 True b'IND' True b'SD'\n",
      "0 True b'IND' True b'MU'\n",
      "0 True b'IND' False b'BI'\n",
      "0 True b'IND' False b'SD'\n",
      "0 True b'IND' False b'MU'\n",
      "0 False b'SNP' True b'BI'\n",
      "0 False b'SNP' True b'SD'\n",
      "0 False b'SNP' True b'MU'\n",
      "0 False b'SNP' False b'BI'\n",
      "0 False b'SNP' False b'SD'\n",
      "0 False b'SNP' False b'MU'\n",
      "0 False b'IND' True b'BI'\n",
      "0 False b'IND' True b'SD'\n",
      "0 False b'IND' True b'MU'\n",
      "0 False b'IND' False b'BI'\n",
      "0 False b'IND' False b'SD'\n",
      "0 False b'IND' False b'MU'\n",
      "30 True b'SNP' True b'BI'\n",
      "30 True b'SNP' True b'SD'\n",
      "30 True b'SNP' True b'MU'\n",
      "30 True b'SNP' False b'BI'\n",
      "30 True b'SNP' False b'SD'\n",
      "30 True b'SNP' False b'MU'\n",
      "30 True b'IND' True b'BI'\n",
      "30 True b'IND' True b'SD'\n",
      "30 True b'IND' True b'MU'\n",
      "30 True b'IND' False b'BI'\n",
      "30 True b'IND' False b'SD'\n",
      "30 True b'IND' False b'MU'\n",
      "30 False b'SNP' True b'BI'\n",
      "30 False b'SNP' True b'SD'\n",
      "30 False b'SNP' True b'MU'\n",
      "30 False b'SNP' False b'BI'\n",
      "30 False b'SNP' False b'SD'\n",
      "30 False b'SNP' False b'MU'\n",
      "30 False b'IND' True b'BI'\n",
      "30 False b'IND' True b'SD'\n",
      "30 False b'IND' True b'MU'\n",
      "30 False b'IND' False b'BI'\n",
      "30 False b'IND' False b'SD'\n",
      "30 False b'IND' False b'MU'\n",
      "99 True b'SNP' True b'BI'\n",
      "99 True b'SNP' True b'SD'\n",
      "99 True b'SNP' True b'MU'\n",
      "99 True b'SNP' False b'BI'\n",
      "99 True b'SNP' False b'SD'\n",
      "99 True b'SNP' False b'MU'\n",
      "99 True b'IND' True b'BI'\n",
      "99 True b'IND' True b'SD'\n",
      "99 True b'IND' True b'MU'\n",
      "99 True b'IND' False b'BI'\n",
      "99 True b'IND' False b'SD'\n",
      "99 True b'IND' False b'MU'\n",
      "99 False b'SNP' True b'BI'\n",
      "99 False b'SNP' True b'SD'\n",
      "99 False b'SNP' True b'MU'\n",
      "99 False b'SNP' False b'BI'\n",
      "99 False b'SNP' False b'SD'\n",
      "99 False b'SNP' False b'MU'\n",
      "99 False b'IND' True b'BI'\n",
      "99 False b'IND' True b'SD'\n",
      "99 False b'IND' True b'MU'\n",
      "99 False b'IND' False b'BI'\n",
      "99 False b'IND' False b'SD'\n",
      "99 False b'IND' False b'MU'\n"
     ]
    }
   ],
   "source": [
    "results_list = list()\n",
    "# GQ_thresholds = [30, 99, 0]\n",
    "GQ_thresholds = [0, 30, 99]\n",
    "variant_types = [b'SNP', b'IND']\n",
    "# variant_types = [b'SNP', b'IND', b'STR']\n",
    "multiallelics = [b'BI', b'SD', b'MU']\n",
    "codings = [True, False]\n",
    "filter_passes = [True, False]\n",
    "\n",
    "for GQ_threshold in GQ_thresholds:\n",
    "    for filter_pass in filter_passes:\n",
    "        for variant_type in variant_types:\n",
    "            for coding in codings:\n",
    "                for multiallelic in multiallelics:\n",
    "                    print(GQ_threshold, filter_pass, variant_type, coding, multiallelic)\n",
    "                    ix = (\n",
    "                        (variants['VARIANT_TYPE'] == variant_type) &\n",
    "                        (variants['MULTIALLELIC'] == multiallelic) &\n",
    "                        (variants['CDS'] == coding) &\n",
    "                        (variants['FILTER_PASS'] == filter_pass)\n",
    "                    )\n",
    "                    number_of_alleles = np.sum(ix)\n",
    "                    number_of_sites = len(np.unique(variants[['CHROM', 'POS']][ix]))\n",
    "                    mean_nraf = np.sum(variants['AC'][ix]) / np.sum(variants['AN'][ix])\n",
    "                    genotype_quality_results = list(genotype_concordance_gatk(ix=ix, GQ_threshold=GQ_threshold))\n",
    "#                     sites_ix = (\n",
    "#                         (site_variants['VARIANT_TYPE'] == variant_type) &\n",
    "#                         (site_variants['MULTIALLELIC'] == multiallelic) &\n",
    "#                         (site_variants['CDS'] == coding) &\n",
    "#                         (site_variants['FILTER_PASS'] == filter_pass)\n",
    "#                     )\n",
    "#                     num_sites = np.sum(sites_ix)\n",
    "                    results_list.append(\n",
    "                        [GQ_threshold, filter_pass, variant_type, coding, multiallelic, number_of_sites, number_of_alleles, mean_nraf] +\n",
    "                        genotype_quality_results\n",
    "                    )\n",
    "\n",
    "# print(results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], \n",
       "      dtype=[('CHROM', 'S12'), ('POS', '<i4')])"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity check. Previously this was showing 2 variants, which was due to a bug\n",
    "variant_type = b'SNP'\n",
    "multiallelic = b'BI'\n",
    "coding = False\n",
    "filter_pass = True\n",
    "ix = (\n",
    "                        (variants['VARIANT_TYPE'] == variant_type) &\n",
    "                        (variants['MULTIALLELIC'] == multiallelic) &\n",
    "                        (variants['CDS'] == coding) &\n",
    "                        (variants['FILTER_PASS'] == filter_pass)\n",
    "                    )\n",
    "temp = variants[['CHROM', 'POS']][ix]\n",
    "s = np.sort(temp, axis=None)\n",
    "s[s[1:] == s[:-1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ (b'Pf3D7_01_v3', 514753, b'A', b'C', b'MULTIALLELIC_MIXED', b'SNP', b'INTRAGENIC', 6, 162, False, b'MU', 3.0399999618530273, False, False, False, False, True, False, False, False, False, False, False, False),\n",
       "       (b'Pf3D7_01_v3', 514753, b'A', b'T', b'MULTIALLELIC_MIXED', b'SNP', b'INTRAGENIC', 12, 162, False, b'MU', 3.0399999618530273, False, False, False, False, True, False, False, False, False, False, False, False)], \n",
       "      dtype=[('CHROM', 'S12'), ('POS', '<i4'), ('REF', 'S10'), ('ALT', 'S10'), ('VariantType', 'S40'), ('VARIANT_TYPE', 'S3'), ('SNPEFF_EFFECT', 'S33'), ('AC', '<u2'), ('AN', '<u2'), ('CDS', '?'), ('MULTIALLELIC', 'S2'), ('VQSLOD', '<f4'), ('FILTER_Centromere', '?'), ('FILTER_InternalHypervariable', '?'), ('FILTER_LowQual', '?'), ('FILTER_Low_VQSLOD', '?'), ('FILTER_PASS', '?'), ('FILTER_SubtelomericHypervariable', '?'), ('FILTER_SubtelomericRepeat', '?'), ('FILTER_VQSRTrancheINDEL99.00to99.90', '?'), ('FILTER_VQSRTrancheINDEL99.90to100.00', '?'), ('FILTER_VQSRTrancheINDEL99.90to100.00+', '?'), ('FILTER_VQSRTrancheSNP99.90to100.00', '?'), ('FILTER_VQSRTrancheSNP99.90to100.00+', '?')])"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sanity check. Previously this was 1 of the two variants shown above and multiallelic was b'BI' not b'MU'\n",
    "variants[(variants['CHROM']==b'Pf3D7_01_v3') & (variants['POS']==514753)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>GQ threshold</th>\n",
       "<th>PASS</th>\n",
       "<th>Type</th>\n",
       "<th>Coding</th>\n",
       "<th>Multiallelic</th>\n",
       "<th>Variants</th>\n",
       "<th>Alleles</th>\n",
       "<th>Mean NRAF</th>\n",
       "<th>Missingness</th>\n",
       "<th>Heterozygosity</th>\n",
       "<th>Discordance</th>\n",
       "<th>MER</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>11745</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.00862807447372</td>\n",
       "<td style='text-align: right'>0.00492871478093</td>\n",
       "<td style='text-align: right'>4.79772779612e-05</td>\n",
       "<td style='text-align: right'>8.81303571175e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>2809</td>\n",
       "<td>2809</td>\n",
       "<td style='text-align: right'>0.145712725503</td>\n",
       "<td style='text-align: right'>0.0819850189987</td>\n",
       "<td style='text-align: right'>0.0206333868542</td>\n",
       "<td style='text-align: right'>0.00830579619541</td>\n",
       "<td style='text-align: right'>0.00890223345976</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>7042</td>\n",
       "<td>7197</td>\n",
       "<td style='text-align: right'>0.238426438289</td>\n",
       "<td style='text-align: right'>0.083996166203</td>\n",
       "<td style='text-align: right'>0.0208136610209</td>\n",
       "<td style='text-align: right'>0.00689282099606</td>\n",
       "<td style='text-align: right'>0.00500151875617</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>9538</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.0169654357864</td>\n",
       "<td style='text-align: right'>0.00844527368507</td>\n",
       "<td style='text-align: right'>0.000293717443819</td>\n",
       "<td style='text-align: right'>0.000267352630894</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>15682</td>\n",
       "<td>15682</td>\n",
       "<td style='text-align: right'>0.0612402838683</td>\n",
       "<td style='text-align: right'>0.0377275128901</td>\n",
       "<td style='text-align: right'>0.0226445762593</td>\n",
       "<td style='text-align: right'>0.00417457895003</td>\n",
       "<td style='text-align: right'>0.00292394846922</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10987</td>\n",
       "<td>11192</td>\n",
       "<td style='text-align: right'>0.153550008888</td>\n",
       "<td style='text-align: right'>0.0562719726919</td>\n",
       "<td style='text-align: right'>0.0304782205949</td>\n",
       "<td style='text-align: right'>0.00680720245938</td>\n",
       "<td style='text-align: right'>0.00441946192542</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>287</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.0143994880182</td>\n",
       "<td style='text-align: right'>0.0136528478987</td>\n",
       "<td style='text-align: right'>0.00120724346076</td>\n",
       "<td style='text-align: right'>0.000957510472771</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>900</td>\n",
       "<td>900</td>\n",
       "<td style='text-align: right'>0.122218278616</td>\n",
       "<td style='text-align: right'>0.0416666666667</td>\n",
       "<td style='text-align: right'>0.0157823129252</td>\n",
       "<td style='text-align: right'>0.0113162596784</td>\n",
       "<td style='text-align: right'>0.010324423587</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10963</td>\n",
       "<td>16451</td>\n",
       "<td style='text-align: right'>0.199129503224</td>\n",
       "<td style='text-align: right'>0.0325015909956</td>\n",
       "<td style='text-align: right'>0.021337329534</td>\n",
       "<td style='text-align: right'>0.00943164980701</td>\n",
       "<td style='text-align: right'>0.00649421668869</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1929</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.0223971392601</td>\n",
       "<td style='text-align: right'>0.0139598607717</td>\n",
       "<td style='text-align: right'>0.00252745599519</td>\n",
       "<td style='text-align: right'>0.00170539592177</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>10549</td>\n",
       "<td>10549</td>\n",
       "<td style='text-align: right'>0.0522823401716</td>\n",
       "<td style='text-align: right'>0.0410842695216</td>\n",
       "<td style='text-align: right'>0.0211887769612</td>\n",
       "<td style='text-align: right'>0.00538105224345</td>\n",
       "<td style='text-align: right'>0.00365308216791</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78928</td>\n",
       "<td>175422</td>\n",
       "<td style='text-align: right'>0.131376167181</td>\n",
       "<td style='text-align: right'>0.0296180824828</td>\n",
       "<td style='text-align: right'>0.0431645415289</td>\n",
       "<td style='text-align: right'>0.0106921686689</td>\n",
       "<td style='text-align: right'>0.00598589043912</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>14316</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.177776684857</td>\n",
       "<td style='text-align: right'>0.0873205946251</td>\n",
       "<td style='text-align: right'>0.00586705855754</td>\n",
       "<td style='text-align: right'>0.00552638686489</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8721</td>\n",
       "<td>8721</td>\n",
       "<td style='text-align: right'>0.109055550703</td>\n",
       "<td style='text-align: right'>0.236491087663</td>\n",
       "<td style='text-align: right'>0.0637015039934</td>\n",
       "<td style='text-align: right'>0.0071663153767</td>\n",
       "<td style='text-align: right'>0.00663005962551</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>86808</td>\n",
       "<td>100608</td>\n",
       "<td style='text-align: right'>0.129334907691</td>\n",
       "<td style='text-align: right'>0.348730737524</td>\n",
       "<td style='text-align: right'>0.0551924908799</td>\n",
       "<td style='text-align: right'>0.0104583340852</td>\n",
       "<td style='text-align: right'>0.0104969982858</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>43798</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.143722432578</td>\n",
       "<td style='text-align: right'>0.0705399836541</td>\n",
       "<td style='text-align: right'>0.00301074395821</td>\n",
       "<td style='text-align: right'>0.00265698854197</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>39053</td>\n",
       "<td>39053</td>\n",
       "<td style='text-align: right'>0.0662007904141</td>\n",
       "<td style='text-align: right'>0.113871154689</td>\n",
       "<td style='text-align: right'>0.0800847827416</td>\n",
       "<td style='text-align: right'>0.00409691722209</td>\n",
       "<td style='text-align: right'>0.00250512552843</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>131903</td>\n",
       "<td>150449</td>\n",
       "<td style='text-align: right'>0.0893861762782</td>\n",
       "<td style='text-align: right'>0.194831566084</td>\n",
       "<td style='text-align: right'>0.0763133374507</td>\n",
       "<td style='text-align: right'>0.00886872602153</td>\n",
       "<td style='text-align: right'>0.00600770645994</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1105</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.12075907286</td>\n",
       "<td style='text-align: right'>0.0397820666728</td>\n",
       "<td style='text-align: right'>0.00503619767076</td>\n",
       "<td style='text-align: right'>0.00305282351641</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>1266</td>\n",
       "<td>1266</td>\n",
       "<td style='text-align: right'>0.0579327337923</td>\n",
       "<td style='text-align: right'>0.159315859045</td>\n",
       "<td style='text-align: right'>0.0375600477158</td>\n",
       "<td style='text-align: right'>0.00958958844683</td>\n",
       "<td style='text-align: right'>0.00590672190298</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>46417</td>\n",
       "<td>68471</td>\n",
       "<td style='text-align: right'>0.0746655169402</td>\n",
       "<td style='text-align: right'>0.322691209358</td>\n",
       "<td style='text-align: right'>0.0354182420146</td>\n",
       "<td style='text-align: right'>0.0120518777997</td>\n",
       "<td style='text-align: right'>0.00859840624752</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>3656</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.140657794847</td>\n",
       "<td style='text-align: right'>0.0381536640915</td>\n",
       "<td style='text-align: right'>0.00494936419706</td>\n",
       "<td style='text-align: right'>0.00299365398497</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8073</td>\n",
       "<td>8073</td>\n",
       "<td style='text-align: right'>0.0319228272062</td>\n",
       "<td style='text-align: right'>0.0830647383442</td>\n",
       "<td style='text-align: right'>0.0348794292894</td>\n",
       "<td style='text-align: right'>0.00354758941703</td>\n",
       "<td style='text-align: right'>0.00198206198384</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78444</td>\n",
       "<td>129526</td>\n",
       "<td style='text-align: right'>0.0522128904109</td>\n",
       "<td style='text-align: right'>0.126851294847</td>\n",
       "<td style='text-align: right'>0.0527332468432</td>\n",
       "<td style='text-align: right'>0.00847241081836</td>\n",
       "<td style='text-align: right'>0.0044919621873</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>11745</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.0171849071685</td>\n",
       "<td style='text-align: right'>0.00424062345245</td>\n",
       "<td style='text-align: right'>1.44995480974e-05</td>\n",
       "<td style='text-align: right'>3.16620868577e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>2809</td>\n",
       "<td>2809</td>\n",
       "<td style='text-align: right'>0.145712725503</td>\n",
       "<td style='text-align: right'>0.206831540021</td>\n",
       "<td style='text-align: right'>0.0154387137553</td>\n",
       "<td style='text-align: right'>0.00243408747342</td>\n",
       "<td style='text-align: right'>0.00533202870997</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>7042</td>\n",
       "<td>7197</td>\n",
       "<td style='text-align: right'>0.238426438289</td>\n",
       "<td style='text-align: right'>0.205946071634</td>\n",
       "<td style='text-align: right'>0.0167927112487</td>\n",
       "<td style='text-align: right'>0.00228571428571</td>\n",
       "<td style='text-align: right'>0.00144351872123</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>9538</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.0956902786277</td>\n",
       "<td style='text-align: right'>0.00655808559532</td>\n",
       "<td style='text-align: right'>0.000103706200334</td>\n",
       "<td style='text-align: right'>0.000171851075548</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>15682</td>\n",
       "<td>15682</td>\n",
       "<td style='text-align: right'>0.0612402838683</td>\n",
       "<td style='text-align: right'>0.289433615558</td>\n",
       "<td style='text-align: right'>0.0132180662088</td>\n",
       "<td style='text-align: right'>0.00190649727523</td>\n",
       "<td style='text-align: right'>0.0013242818144</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10987</td>\n",
       "<td>11192</td>\n",
       "<td style='text-align: right'>0.153550008888</td>\n",
       "<td style='text-align: right'>0.327105002115</td>\n",
       "<td style='text-align: right'>0.0210773730507</td>\n",
       "<td style='text-align: right'>0.00263120644146</td>\n",
       "<td style='text-align: right'>0.000933901071212</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>287</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.0622200099552</td>\n",
       "<td style='text-align: right'>0.0104529616725</td>\n",
       "<td style='text-align: right'>0.000859475719811</td>\n",
       "<td style='text-align: right'>0.000514822600712</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>900</td>\n",
       "<td>900</td>\n",
       "<td style='text-align: right'>0.122218278616</td>\n",
       "<td style='text-align: right'>0.133922902494</td>\n",
       "<td style='text-align: right'>0.0100340136054</td>\n",
       "<td style='text-align: right'>0.00648498979889</td>\n",
       "<td style='text-align: right'>0.0058160256618</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10963</td>\n",
       "<td>16451</td>\n",
       "<td style='text-align: right'>0.199129503224</td>\n",
       "<td style='text-align: right'>0.159258354123</td>\n",
       "<td style='text-align: right'>0.0131292806467</td>\n",
       "<td style='text-align: right'>0.00488467454785</td>\n",
       "<td style='text-align: right'>0.00355168479147</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1929</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.153484410872</td>\n",
       "<td style='text-align: right'>0.00854836491362</td>\n",
       "<td style='text-align: right'>0.00048666875239</td>\n",
       "<td style='text-align: right'>0.000490265324946</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>10549</td>\n",
       "<td>10549</td>\n",
       "<td style='text-align: right'>0.0522823401716</td>\n",
       "<td style='text-align: right'>0.289873689546</td>\n",
       "<td style='text-align: right'>0.0114170798664</td>\n",
       "<td style='text-align: right'>0.00196753974513</td>\n",
       "<td style='text-align: right'>0.00174657194932</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78928</td>\n",
       "<td>175422</td>\n",
       "<td style='text-align: right'>0.131376167181</td>\n",
       "<td style='text-align: right'>0.391848961769</td>\n",
       "<td style='text-align: right'>0.0218024104672</td>\n",
       "<td style='text-align: right'>0.00199755838008</td>\n",
       "<td style='text-align: right'>0.000786649401939</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>14316</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.418956098785</td>\n",
       "<td style='text-align: right'>0.0801443796295</td>\n",
       "<td style='text-align: right'>0.000174951888231</td>\n",
       "<td style='text-align: right'>0.000302445359688</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8721</td>\n",
       "<td>8721</td>\n",
       "<td style='text-align: right'>0.109055550703</td>\n",
       "<td style='text-align: right'>0.493144626272</td>\n",
       "<td style='text-align: right'>0.0574159488357</td>\n",
       "<td style='text-align: right'>0.000655188176139</td>\n",
       "<td style='text-align: right'>0.000880151693258</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>86808</td>\n",
       "<td>100608</td>\n",
       "<td style='text-align: right'>0.129334907691</td>\n",
       "<td style='text-align: right'>0.599025678974</td>\n",
       "<td style='text-align: right'>0.0498340497936</td>\n",
       "<td style='text-align: right'>0.000521744569001</td>\n",
       "<td style='text-align: right'>0.00113653337214</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>43798</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.4721376244</td>\n",
       "<td style='text-align: right'>0.0615634764797</td>\n",
       "<td style='text-align: right'>0.000104116157935</td>\n",
       "<td style='text-align: right'>0.000271793657653</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>39053</td>\n",
       "<td>39053</td>\n",
       "<td style='text-align: right'>0.0662007904141</td>\n",
       "<td style='text-align: right'>0.515960779621</td>\n",
       "<td style='text-align: right'>0.0677303005805</td>\n",
       "<td style='text-align: right'>0.000669146526465</td>\n",
       "<td style='text-align: right'>0.000625177082178</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>131903</td>\n",
       "<td>150449</td>\n",
       "<td style='text-align: right'>0.0893861762782</td>\n",
       "<td style='text-align: right'>0.527251081491</td>\n",
       "<td style='text-align: right'>0.0679427471591</td>\n",
       "<td style='text-align: right'>0.000396924554514</td>\n",
       "<td style='text-align: right'>0.000542484411687</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1105</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.25597008034</td>\n",
       "<td style='text-align: right'>0.0347215809401</td>\n",
       "<td style='text-align: right'>0.00115313653137</td>\n",
       "<td style='text-align: right'>0.000593781488412</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>1266</td>\n",
       "<td>1266</td>\n",
       "<td style='text-align: right'>0.0579327337923</td>\n",
       "<td style='text-align: right'>0.323282393526</td>\n",
       "<td style='text-align: right'>0.032828771319</td>\n",
       "<td style='text-align: right'>0.00271400278735</td>\n",
       "<td style='text-align: right'>0.00277750988523</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>46417</td>\n",
       "<td>68471</td>\n",
       "<td style='text-align: right'>0.0746655169402</td>\n",
       "<td style='text-align: right'>0.567769641192</td>\n",
       "<td style='text-align: right'>0.0302743989039</td>\n",
       "<td style='text-align: right'>0.00289627004356</td>\n",
       "<td style='text-align: right'>0.00250324813386</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>3656</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.356978185147</td>\n",
       "<td style='text-align: right'>0.0326859956236</td>\n",
       "<td style='text-align: right'>0.000961027793935</td>\n",
       "<td style='text-align: right'>0.000780938964261</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8073</td>\n",
       "<td>8073</td>\n",
       "<td style='text-align: right'>0.0319228272062</td>\n",
       "<td style='text-align: right'>0.303739853429</td>\n",
       "<td style='text-align: right'>0.0269454493057</td>\n",
       "<td style='text-align: right'>0.000551188914489</td>\n",
       "<td style='text-align: right'>0.000549264699037</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78444</td>\n",
       "<td>129526</td>\n",
       "<td style='text-align: right'>0.0522128904109</td>\n",
       "<td style='text-align: right'>0.498904797934</td>\n",
       "<td style='text-align: right'>0.0390012311767</td>\n",
       "<td style='text-align: right'>0.00134013069728</td>\n",
       "<td style='text-align: right'>0.000820050900559</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>11745</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.0523566259198</td>\n",
       "<td style='text-align: right'>0.00324497615138</td>\n",
       "<td style='text-align: right'>1.48126203525e-05</td>\n",
       "<td style='text-align: right'>3.11939640686e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>2809</td>\n",
       "<td>2809</td>\n",
       "<td style='text-align: right'>0.145712725503</td>\n",
       "<td style='text-align: right'>0.317590688821</td>\n",
       "<td style='text-align: right'>0.012492643907</td>\n",
       "<td style='text-align: right'>0.000884277545246</td>\n",
       "<td style='text-align: right'>0.00327456701223</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>7042</td>\n",
       "<td>7197</td>\n",
       "<td style='text-align: right'>0.238426438289</td>\n",
       "<td style='text-align: right'>0.28859105126</td>\n",
       "<td style='text-align: right'>0.0138833357436</td>\n",
       "<td style='text-align: right'>0.000978846049269</td>\n",
       "<td style='text-align: right'>0.00059951366555</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>9538</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.265373522024</td>\n",
       "<td style='text-align: right'>0.00424082402934</td>\n",
       "<td style='text-align: right'>0.000115772325453</td>\n",
       "<td style='text-align: right'>0.000145753811251</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>15682</td>\n",
       "<td>15682</td>\n",
       "<td style='text-align: right'>0.0612402838683</td>\n",
       "<td style='text-align: right'>0.62764211666</td>\n",
       "<td style='text-align: right'>0.00583536564734</td>\n",
       "<td style='text-align: right'>0.000847407231208</td>\n",
       "<td style='text-align: right'>0.00106409959531</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10987</td>\n",
       "<td>11192</td>\n",
       "<td style='text-align: right'>0.153550008888</td>\n",
       "<td style='text-align: right'>0.594294758647</td>\n",
       "<td style='text-align: right'>0.0121415077825</td>\n",
       "<td style='text-align: right'>0.000828025477707</td>\n",
       "<td style='text-align: right'>0.000114298154901</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>287</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.133470809927</td>\n",
       "<td style='text-align: right'>0.00764417265164</td>\n",
       "<td style='text-align: right'>0.000468055230517</td>\n",
       "<td style='text-align: right'>0.000381642972999</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>900</td>\n",
       "<td>900</td>\n",
       "<td style='text-align: right'>0.122218278616</td>\n",
       "<td style='text-align: right'>0.268888888889</td>\n",
       "<td style='text-align: right'>0.00707482993197</td>\n",
       "<td style='text-align: right'>0.00344322974951</td>\n",
       "<td style='text-align: right'>0.00348042262275</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>10963</td>\n",
       "<td>16451</td>\n",
       "<td style='text-align: right'>0.199129503224</td>\n",
       "<td style='text-align: right'>0.299100358641</td>\n",
       "<td style='text-align: right'>0.00690982125024</td>\n",
       "<td style='text-align: right'>0.00296920031589</td>\n",
       "<td style='text-align: right'>0.00224726570801</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1929</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.387411263106</td>\n",
       "<td style='text-align: right'>0.00426889262704</td>\n",
       "<td style='text-align: right'>0.00021393806493</td>\n",
       "<td style='text-align: right'>0.00046332539822</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>10549</td>\n",
       "<td>10549</td>\n",
       "<td style='text-align: right'>0.0522823401716</td>\n",
       "<td style='text-align: right'>0.637271934084</td>\n",
       "<td style='text-align: right'>0.00432674728817</td>\n",
       "<td style='text-align: right'>0.000749267761052</td>\n",
       "<td style='text-align: right'>0.00115933032699</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78928</td>\n",
       "<td>175422</td>\n",
       "<td style='text-align: right'>0.131376167181</td>\n",
       "<td style='text-align: right'>0.718223739884</td>\n",
       "<td style='text-align: right'>0.0060643849153</td>\n",
       "<td style='text-align: right'>0.00055329650545</td>\n",
       "<td style='text-align: right'>0.000295026415861</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>14316</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.580954804386</td>\n",
       "<td style='text-align: right'>0.0640691733525</td>\n",
       "<td style='text-align: right'>1.13163135977e-05</td>\n",
       "<td style='text-align: right'>9.99510239982e-06</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8721</td>\n",
       "<td>8721</td>\n",
       "<td style='text-align: right'>0.109055550703</td>\n",
       "<td style='text-align: right'>0.645492114975</td>\n",
       "<td style='text-align: right'>0.0446248674909</td>\n",
       "<td style='text-align: right'>6.3041102799e-05</td>\n",
       "<td style='text-align: right'>0.000355250922085</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>86808</td>\n",
       "<td>100608</td>\n",
       "<td style='text-align: right'>0.129334907691</td>\n",
       "<td style='text-align: right'>0.736963445922</td>\n",
       "<td style='text-align: right'>0.0379021062146</td>\n",
       "<td style='text-align: right'>4.63815438561e-05</td>\n",
       "<td style='text-align: right'>0.000261382646011</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>43798</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.672177976629</td>\n",
       "<td style='text-align: right'>0.0452604303057</td>\n",
       "<td style='text-align: right'>3.95198340167e-05</td>\n",
       "<td style='text-align: right'>6.23866123321e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>39053</td>\n",
       "<td>39053</td>\n",
       "<td style='text-align: right'>0.0662007904141</td>\n",
       "<td style='text-align: right'>0.738194614645</td>\n",
       "<td style='text-align: right'>0.046418080714</td>\n",
       "<td style='text-align: right'>0.000189890899047</td>\n",
       "<td style='text-align: right'>0.000345117889133</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>131903</td>\n",
       "<td>150449</td>\n",
       "<td style='text-align: right'>0.0893861762782</td>\n",
       "<td style='text-align: right'>0.706769573146</td>\n",
       "<td style='text-align: right'>0.0498172070243</td>\n",
       "<td style='text-align: right'>1.73814937656e-05</td>\n",
       "<td style='text-align: right'>5.5133266694e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>1105</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.406344076092</td>\n",
       "<td style='text-align: right'>0.0277126235109</td>\n",
       "<td style='text-align: right'>0.000408329930584</td>\n",
       "<td style='text-align: right'>0.000146448620942</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>1266</td>\n",
       "<td>1266</td>\n",
       "<td style='text-align: right'>0.0579327337923</td>\n",
       "<td style='text-align: right'>0.486684721282</td>\n",
       "<td style='text-align: right'>0.0261308314795</td>\n",
       "<td style='text-align: right'>0.00136692052334</td>\n",
       "<td style='text-align: right'>0.0016072385261</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>46417</td>\n",
       "<td>68471</td>\n",
       "<td style='text-align: right'>0.0746655169402</td>\n",
       "<td style='text-align: right'>0.724606931759</td>\n",
       "<td style='text-align: right'>0.0219334626696</td>\n",
       "<td style='text-align: right'>0.00056841929282</td>\n",
       "<td style='text-align: right'>0.000764796913899</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>3656</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.586773768588</td>\n",
       "<td style='text-align: right'>0.0243212164516</td>\n",
       "<td style='text-align: right'>0.000639386189258</td>\n",
       "<td style='text-align: right'>0.000468498182227</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>8073</td>\n",
       "<td>8073</td>\n",
       "<td style='text-align: right'>0.0319228272062</td>\n",
       "<td style='text-align: right'>0.586032554977</td>\n",
       "<td style='text-align: right'>0.0167590633429</td>\n",
       "<td style='text-align: right'>0.000195541650372</td>\n",
       "<td style='text-align: right'>0.000312514361873</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>78444</td>\n",
       "<td>129526</td>\n",
       "<td style='text-align: right'>0.0522128904109</td>\n",
       "<td style='text-align: right'>0.731725676698</td>\n",
       "<td style='text-align: right'>0.0235831620915</td>\n",
       "<td style='text-align: right'>0.000166546869795</td>\n",
       "<td style='text-align: right'>0.000125420060227</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "headers = ['GQ threshold', 'PASS', 'Type', 'Coding', 'Multiallelic', 'Variants', 'Alleles', 'Mean NRAF', 'Missingness',\n",
    "           'Heterozygosity', 'Discordance', 'MER']\n",
    "etl.wrap(results_list).pushheader(headers).displayall()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "624919"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(etl.wrap(results_list).pushheader(headers).values('Variants').array())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/genotype_quality.xlsx'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# etl.wrap(results_list).pushheader(headers).convert('Alleles', int).toxlsx(results_table_fn)\n",
    "etl.wrap(results_list).pushheader(headers).cutout('Alleles').cutout('Mean NRAF').toxlsx(results_table_fn)\n",
    "results_table_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# variant_type_header_2_fn = \"%s/vcf/VARIANT_TYPE_2.hdr\" % (output_dir)\n",
    "# fo=open(variant_type_header_2_fn, 'w')\n",
    "# print('##INFO=<ID=VARIANT_TYPE,Number=1,Type=String,Description=\"SNP or INDEL\">', file=fo)\n",
    "# fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_variant_counts_vcf(input_vcf_fn=sites_only_vcf_fn, region='Pf3D7_14_v3:1000000-1100000',\n",
    "                        output_vcf_fn=None, BCFTOOLS=BCFTOOLS, rewrite=False):\n",
    "    if output_vcf_fn is None:\n",
    "        if region is None:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir)\n",
    "        else:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_%s.sites.analysis.vcf.gz\" % (output_dir, region)\n",
    "    intermediate_fns = collections.OrderedDict()\n",
    "    for intermediate_file in ['multiallelic', 'multiallelics', 'snps', 'indels', 'triallelic', 'bi_allelic', 'spanning_deletion',\n",
    "                              'triallelic_no_sd', 'biallelic',\n",
    "                              'variant_type', 'analysis']:\n",
    "        intermediate_fns[intermediate_file] = output_vcf_fn.replace('analysis', intermediate_file)\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tMU\\n' --include 'N_ALT>2' {input_vcf_fn} | bgzip -c > {intermediate_fns['multiallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['multiallelic']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['triallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t%ALT\\tSD\\n' --include 'N_ALT=2' {input_vcf_fn} | bgzip -c > {intermediate_fns['triallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['triallelic']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['bi_allelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t%ALT\\tBI\\n' --include 'N_ALT=1' {input_vcf_fn} | bgzip -c > {intermediate_fns['bi_allelic']} && tabix -s1 -b2 -e2 {intermediate_fns['bi_allelic']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['spanning_deletion']):\n",
    "        !zgrep '\\*' {intermediate_fns['triallelic']} | bgzip -c > {intermediate_fns['spanning_deletion']} && tabix -s1 -b2 -e2 {intermediate_fns['spanning_deletion']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['triallelic_no_sd']):\n",
    "        !zgrep -v '\\*' {intermediate_fns['triallelic']} | sed 's/SD/MU/g' | bgzip -c > {intermediate_fns['triallelic_no_sd']} && tabix -s1 -b2 -e2 {intermediate_fns['triallelic_no_sd']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelics']):\n",
    "        !{BCFTOOLS} annotate --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set\\\n",
    "        -a {intermediate_fns['bi_allelic']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "        -h {multiallelic_header_fn} {input_vcf_fn} | \\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['spanning_deletion']} -c CHROM,POS,-,INFO/MULTIALLELIC |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['multiallelic']} -c CHROM,POS,INFO/MULTIALLELIC |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['triallelic_no_sd']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "        -Oz -o {intermediate_fns['multiallelics']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['multiallelics']}\n",
    "                \n",
    "#     if rewrite or not os.path.exists(intermediate_fns['multiallelics']):\n",
    "#         !{BCFTOOLS} annotate -a {intermediate_fns['bi_allelic']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "#         -h {multiallelic_header_fn} {intermediate_fns['nonref']} | \\\n",
    "#         {BCFTOOLS} annotate -a {intermediate_fns['spanning_deletion']} -c CHROM,POS,-,INFO/MULTIALLELIC |\\\n",
    "#         {BCFTOOLS} annotate -a {intermediate_fns['multiallelic']} -c CHROM,POS,INFO/MULTIALLELIC |\\\n",
    "#         {BCFTOOLS} annotate -a {intermediate_fns['triallelic_no_sd']} -c CHROM,POS,-,INFO/MULTIALLELIC \\\n",
    "#         -Oz -o {intermediate_fns['multiallelics']}\n",
    "#         !{BCFTOOLS} index --tbi {intermediate_fns['multiallelics']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['biallelic']):\n",
    "        !{BCFTOOLS} norm -m -any --fasta-ref {GENOME_FN} {intermediate_fns['multiallelics']} | \\\n",
    "        {BCFTOOLS} view --include 'ALT!=\"*\"' -Oz -o {intermediate_fns['biallelic']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['biallelic']}\n",
    "\n",
    "#         {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['biallelic']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['snps']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSNP\\n' --include 'TYPE=\"snp\"' {intermediate_fns['biallelic']} | bgzip -c > {intermediate_fns['snps']} && tabix -s1 -b2 -e2 -f {intermediate_fns['snps']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['indels']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tINDEL\\n' --include 'TYPE!=\"snp\"' {intermediate_fns['biallelic']} | bgzip -c > {intermediate_fns['indels']} && tabix -s1 -b2 -e2 -f {intermediate_fns['indels']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['variant_type']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['snps']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -h {variant_type_header_fn} {intermediate_fns['biallelic']} | \\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['indels']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -Oz -o {intermediate_fns['variant_type']} \n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['variant_type']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['analysis']):\n",
    "        !{BCFTOOLS} annotate -a {cds_gff_fn} -c CHROM,FROM,TO,CDS \\\n",
    "        -h {cds_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['analysis']} {intermediate_fns['variant_type']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['analysis']}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/SNP_INDEL_WG.combined.filtered.sites.vcf.gz'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sites_only_vcf_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# create_variant_counts_vcf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "create_variant_counts_vcf(region=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# def create_variants_npy_2(vcf_fn):\n",
    "#     output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "#     if not os.path.exists(output_dir):\n",
    "#         os.makedirs(output_dir)\n",
    "#     vcfnp.variants(\n",
    "#         vcf_fn,\n",
    "#         fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE',\n",
    "#                 'SNPEFF_EFFECT', 'AC', 'AN', 'CDS', 'MULTIALLELIC',\n",
    "#                 'VQSLOD', 'FILTER'],\n",
    "#         dtypes={\n",
    "#             'REF':                      'a10',\n",
    "#             'ALT':                      'a10',\n",
    "#             'RegionType':               'a25',\n",
    "#             'VariantType':              'a40',\n",
    "#             'VARIANT_TYPE':             'a3',\n",
    "#             'SNPEFF_EFFECT':            'a33',\n",
    "#             'CDS':                      bool,\n",
    "#             'MULTIALLELIC':             bool,\n",
    "#         },\n",
    "#         arities={\n",
    "#             'ALT':   1,\n",
    "#             'AF':    1,\n",
    "#             'AC':    1,\n",
    "#             'ANN':   1,\n",
    "#         },\n",
    "#         fills={\n",
    "#             'VQSLOD': np.nan,\n",
    "#         },\n",
    "#         flatten_filter=True,\n",
    "#         progress=100000,\n",
    "#         verbose=True,\n",
    "#         cache=True,\n",
    "#         cachedir=output_dir\n",
    "#     )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[vcfnp] 2016-07-21 18:08:26.093835 :: caching is enabled\n",
      "[vcfnp] 2016-07-21 18:08:26.094684 :: cache file available\n",
      "[vcfnp] 2016-07-21 18:08:26.095077 :: loading from cache file /lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz.vcfnp_cache/variants.npy\n"
     ]
    }
   ],
   "source": [
    "create_variants_npy(\"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sites_vcf_fn = \"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir)\n",
    "variants_all = np.load(\"%s.vcfnp_cache/variants.npy\" % sites_vcf_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True b'SNP' True b'BI'\n",
      "True b'SNP' True b'SD'\n",
      "True b'SNP' True b'MU'\n",
      "True b'SNP' False b'BI'\n",
      "True b'SNP' False b'SD'\n",
      "True b'SNP' False b'MU'\n",
      "True b'IND' True b'BI'\n",
      "True b'IND' True b'SD'\n",
      "True b'IND' True b'MU'\n",
      "True b'IND' False b'BI'\n",
      "True b'IND' False b'SD'\n",
      "True b'IND' False b'MU'\n",
      "False b'SNP' True b'BI'\n",
      "False b'SNP' True b'SD'\n",
      "False b'SNP' True b'MU'\n",
      "False b'SNP' False b'BI'\n",
      "False b'SNP' False b'SD'\n",
      "False b'SNP' False b'MU'\n",
      "False b'IND' True b'BI'\n",
      "False b'IND' True b'SD'\n",
      "False b'IND' True b'MU'\n",
      "False b'IND' False b'BI'\n",
      "False b'IND' False b'SD'\n",
      "False b'IND' False b'MU'\n",
      "[[True, b'SNP', True, b'BI', 641856, 641856, 0.0077220200341972644], [True, b'SNP', True, b'SD', 64127, 64127, 0.0088083459494743919], [True, b'SNP', True, b'MU', 54191, 107672, 0.0086526424335729175], [True, b'SNP', False, b'BI', 417751, 417751, 0.0082117160465197854], [True, b'SNP', False, b'SD', 131513, 131513, 0.012300331206515698], [True, b'SNP', False, b'MU', 40590, 72951, 0.0098582693489400568], [True, b'IND', True, b'BI', 7310, 7310, 0.010850542187022943], [True, b'IND', True, b'SD', 22336, 22336, 0.0076193744694664622], [True, b'IND', True, b'MU', 86697, 229894, 0.021748752745054355], [True, b'IND', False, b'BI', 67700, 67700, 0.0073883577101844832], [True, b'IND', False, b'SD', 106394, 106394, 0.0093341272824966418], [True, b'IND', False, b'MU', 294768, 803369, 0.032106565508140734], [False, b'SNP', True, b'BI', 369312, 369312, 0.0056312908353769522], [False, b'SNP', True, b'SD', 59590, 59590, 0.019292963026264244], [False, b'SNP', True, b'MU', 87969, 182670, 0.030268711785263996], [False, b'SNP', False, b'BI', 424579, 424579, 0.0092001860773348026], [False, b'SNP', False, b'SD', 177266, 177266, 0.014945157265209731], [False, b'SNP', False, b'MU', 204954, 422095, 0.018675713700732309], [False, b'IND', True, b'BI', 29418, 29418, 0.0030583324924715219], [False, b'IND', True, b'SD', 21779, 21779, 0.0047679110654596903], [False, b'IND', True, b'MU', 316025, 1085570, 0.018954676324932689], [False, b'IND', False, b'BI', 78760, 78760, 0.0027401173723013493], [False, b'IND', False, b'SD', 64935, 64935, 0.0043746221542510207], [False, b'IND', False, b'MU', 442939, 1236125, 0.012417412451261123]]\n"
     ]
    }
   ],
   "source": [
    "counts_list = list()\n",
    "GQ_thresholds = [0, 30, 99]\n",
    "variant_types = [b'SNP', b'IND']\n",
    "# variant_types = [b'SNP', b'IND', b'STR']\n",
    "multiallelics = [b'BI', b'SD', b'MU']\n",
    "codings = [True, False]\n",
    "filter_passes = [True, False]\n",
    "\n",
    "# GQ_thresholds = [30, 99, 0]\n",
    "# variant_types = [b'SNP', b'IND']\n",
    "# multiallelics = [False, True]\n",
    "# codings = [True, False]\n",
    "# filter_passes = [True, False]\n",
    "\n",
    "\n",
    "for filter_pass in filter_passes:\n",
    "    for variant_type in variant_types:\n",
    "        for coding in codings:\n",
    "            for multiallelic in multiallelics:\n",
    "                print(filter_pass, variant_type, coding, multiallelic)\n",
    "                ix = (\n",
    "                    (variants_all['VARIANT_TYPE'] == variant_type) &\n",
    "                    (variants_all['MULTIALLELIC'] == multiallelic) &\n",
    "                    (variants_all['CDS'] == coding) &\n",
    "                    (variants_all['FILTER_PASS'] == filter_pass)\n",
    "                )\n",
    "                number_of_alleles = np.sum(ix)\n",
    "                number_of_sites = len(np.unique(variants_all[['CHROM', 'POS']][ix]))\n",
    "                mean_nraf = np.sum(variants_all['AC'][ix]) / np.sum(variants_all['AN'][ix])\n",
    "#                 number_of_variants = np.sum(ix)\n",
    "                counts_list.append(\n",
    "                    [filter_pass, variant_type, coding, multiallelic, number_of_sites, number_of_alleles, mean_nraf]\n",
    "                )\n",
    "\n",
    "print(counts_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>PASS</th>\n",
       "<th>Type</th>\n",
       "<th>Coding</th>\n",
       "<th>Multiallelic</th>\n",
       "<th>Variants</th>\n",
       "<th>Alleles</th>\n",
       "<th>Mean NRAF</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>641856</td>\n",
       "<td style='text-align: right'>641856</td>\n",
       "<td style='text-align: right'>0.007722020034197264</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>64127</td>\n",
       "<td style='text-align: right'>64127</td>\n",
       "<td style='text-align: right'>0.008808345949474392</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>54191</td>\n",
       "<td style='text-align: right'>107672</td>\n",
       "<td style='text-align: right'>0.008652642433572917</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>417751</td>\n",
       "<td style='text-align: right'>417751</td>\n",
       "<td style='text-align: right'>0.008211716046519785</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>131513</td>\n",
       "<td style='text-align: right'>131513</td>\n",
       "<td style='text-align: right'>0.012300331206515698</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>40590</td>\n",
       "<td style='text-align: right'>72951</td>\n",
       "<td style='text-align: right'>0.009858269348940057</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>7310</td>\n",
       "<td style='text-align: right'>7310</td>\n",
       "<td style='text-align: right'>0.010850542187022943</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>22336</td>\n",
       "<td style='text-align: right'>22336</td>\n",
       "<td style='text-align: right'>0.007619374469466462</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>86697</td>\n",
       "<td style='text-align: right'>229894</td>\n",
       "<td style='text-align: right'>0.021748752745054355</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>67700</td>\n",
       "<td style='text-align: right'>67700</td>\n",
       "<td style='text-align: right'>0.007388357710184483</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>106394</td>\n",
       "<td style='text-align: right'>106394</td>\n",
       "<td style='text-align: right'>0.009334127282496642</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>294768</td>\n",
       "<td style='text-align: right'>803369</td>\n",
       "<td style='text-align: right'>0.032106565508140734</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>369312</td>\n",
       "<td style='text-align: right'>369312</td>\n",
       "<td style='text-align: right'>0.005631290835376952</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>59590</td>\n",
       "<td style='text-align: right'>59590</td>\n",
       "<td style='text-align: right'>0.019292963026264244</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>87969</td>\n",
       "<td style='text-align: right'>182670</td>\n",
       "<td style='text-align: right'>0.030268711785263996</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>424579</td>\n",
       "<td style='text-align: right'>424579</td>\n",
       "<td style='text-align: right'>0.009200186077334803</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>177266</td>\n",
       "<td style='text-align: right'>177266</td>\n",
       "<td style='text-align: right'>0.014945157265209731</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>204954</td>\n",
       "<td style='text-align: right'>422095</td>\n",
       "<td style='text-align: right'>0.01867571370073231</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>29418</td>\n",
       "<td style='text-align: right'>29418</td>\n",
       "<td style='text-align: right'>0.003058332492471522</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>21779</td>\n",
       "<td style='text-align: right'>21779</td>\n",
       "<td style='text-align: right'>0.00476791106545969</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>316025</td>\n",
       "<td style='text-align: right'>1085570</td>\n",
       "<td style='text-align: right'>0.01895467632493269</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'BI'</td>\n",
       "<td style='text-align: right'>78760</td>\n",
       "<td style='text-align: right'>78760</td>\n",
       "<td style='text-align: right'>0.0027401173723013493</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'SD'</td>\n",
       "<td style='text-align: right'>64935</td>\n",
       "<td style='text-align: right'>64935</td>\n",
       "<td style='text-align: right'>0.004374622154251021</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>b'MU'</td>\n",
       "<td style='text-align: right'>442939</td>\n",
       "<td style='text-align: right'>1236125</td>\n",
       "<td style='text-align: right'>0.012417412451261123</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "headers = ['PASS', 'Type', 'Coding', 'Multiallelic', 'Variants', 'Alleles', 'Mean NRAF']\n",
    "(etl\n",
    " .wrap(counts_list)\n",
    " .pushheader(headers)\n",
    " .convert('Mean NRAF', float)\n",
    " .convert('Variants', int)\n",
    " .convert('Alleles', int)\n",
    " .displayall()\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160720_mendelian_error_duplicate_concordance/variant_counts.xlsx'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(etl\n",
    " .wrap(counts_list)\n",
    " .pushheader(headers)\n",
    " .convert('Mean NRAF', float)\n",
    " .convert('Variants', int)\n",
    " .convert('Alleles', int)\n",
    " .toxlsx(counts_table_fn)\n",
    ")\n",
    "# etl.wrap(counts_list).pushheader(headers).convertnumbers().toxlsx(counts_table_fn)\n",
    "counts_table_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1935195"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(etl\n",
    "       .fromtsv('/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/passcounts')\n",
    "       .pushheader(('CHROM', 'count'))\n",
    "       .convertnumbers()\n",
    "       .values('count')\n",
    "       .array()\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2+2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
