{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) \n",
      "[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]\n",
      "numpy 1.11.2\n",
      "scipy 0.18.1\n",
      "pandas 0.19.0\n",
      "numexpr 2.6.1\n",
      "pysam 0.8.4\n",
      "petl 1.1.0\n",
      "petlx 1.0.3\n",
      "vcf 0.6.8\n",
      "vcfnp 2.2.0\n",
      "h5py 2.6.0\n",
      "tables 3.3.0\n"
     ]
    }
   ],
   "source": [
    "%run _standard_imports.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "temp_dir = '/lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test'\n",
    "!mkdir -p {temp_dir}\n",
    "\n",
    "release_vcf_fn = '/nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_01_v3.final.vcf.gz'\n",
    "# release_vcf_fn = '/nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf_M76611.final.vcf.gz'\n",
    "# release_vcf_fn = '/nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_API_v3.final.vcf.gz'\n",
    "\n",
    "bam1_fn = '/nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0388_C.bam'\n",
    "bam2_fn = '/nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0327_C.bam'\n",
    "\n",
    "bam_release_1_fn = '/lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam'\n",
    "bam_release_2_fn = '/lustre/scratch116/malaria/pfalciparum/output/4/2/a/9/36832/4_bam_mark_duplicates_v2/pe.1.markdup.bam'\n",
    "\n",
    "genome_fn = '/lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta'\n",
    "\n",
    "GenomeAnalysisTK=\"/software/jre1.7.0_25/bin/java -Xmx2G -jar /nfs/team112_internal/production/tools/bin/gatk/GenomeAnalysisTK-3.5/GenomeAnalysisTK.jar\"\n",
    "GenomeAnalysisTK=\"/software/java/bin/java -Xmx2G -jar /nfs/team112_internal/production/tools/bin/gatk/GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!cp {release_vcf_fn} {temp_dir}/\n",
    "!gunzip {temp_dir}/Pf_60_Pf3D7_API_v3.final.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gzip: /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/Pf_60_Pf3D7_01_v3.final.vcf.gz: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!cp {release_vcf_fn} {temp_dir}/\n",
    "!gunzip {temp_dir}/Pf_60_Pf3D7_01_v3.final.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  11:37:07,885 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  11:37:07,888 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.5-0-g36282e4, Compiled 2015/11/25 04:03:56 \n",
      "INFO  11:37:07,889 HelpFormatter - Copyright (c) 2010 The Broad Institute \n",
      "INFO  11:37:07,889 HelpFormatter - For support and documentation go to http://www.broadinstitute.org/gatk \n",
      "INFO  11:37:07,895 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0388_C.bam -I /nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0327_C.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_01_v3.final.vcf.gz -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test.vcf \n",
      "INFO  11:37:07,899 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.7.0_25-b15. \n",
      "INFO  11:37:07,899 HelpFormatter - Date/Time: 2017/02/03 11:37:07 \n",
      "INFO  11:37:07,900 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  11:37:07,901 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  11:37:08,355 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  11:37:08,576 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  11:37:08,588 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  11:37:08,755 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.14 \n",
      "INFO  11:37:08,828 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "WARN  11:37:09,007 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  11:37:09,624 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  11:37:09,720 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  11:37:09,721 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  11:37:09,722 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  11:37:09,723 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  11:37:09,724 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  11:37:09,903 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  11:37:09,911 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  11:37:10,108 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "INFO  11:37:39,729 ProgressMeter - Pf3D7_01_v3:2201              0.0    30.0 s           49.6 w        0.0%    88.4 h      88.4 h \n",
      "INFO  11:38:39,733 ProgressMeter - Pf3D7_01_v3:7003              0.0    90.0 s          148.8 w        0.0%    83.3 h      83.3 h \n",
      "INFO  11:39:39,736 ProgressMeter - Pf3D7_01_v3:11097              0.0     2.5 m          248.0 w        0.0%    87.6 h      87.6 h \n",
      "INFO  11:40:39,739 ProgressMeter - Pf3D7_01_v3:14794              0.0     3.5 m          347.3 w        0.1%    92.0 h      92.0 h \n",
      "INFO  11:41:39,742 ProgressMeter - Pf3D7_01_v3:18483              0.0     4.5 m          446.5 w        0.1%    94.7 h      94.6 h \n",
      "INFO  11:42:39,754 ProgressMeter - Pf3D7_01_v3:22375              0.0     5.5 m          545.7 w        0.1%    95.6 h      95.5 h \n",
      "INFO  11:43:39,758 ProgressMeter - Pf3D7_01_v3:26960              0.0     6.5 m          644.9 w        0.1%    93.8 h      93.7 h \n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR stack trace \n",
      "java.lang.IllegalArgumentException: Unexpected base in allele bases 'GTGCACCTATGGTGCACTTTAAAAATAGTATCTGTAAACGAAGTGTTAGATAAACACGGCTTCGTGGTAATACCACATATATAGTGATACCACACATGTGATGTGACCCCACATTTACTGCATAGTGGTGCAACTTATTATATTGTGCATGGTGGTGCGACTTA*ATATTGGTGCGAATTTCTATTTTCGTGCATATTGGTGCGCTTTATAAAAAATTTTAGGATCTAGGTCTTGTCTCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTTTTTTAAATCTTAAAAATAGGAAAAACAAATGTACGTTTTAAAAAAAAGAAAAAAAAAAGAAAAAAAAAAAAAAATTTTTTTTTCTATGTCTTAAAAAGAGAAAAAGGAAATGTACATTTTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTTTTTGTCTCAAAAAAAAAAAATAGTG'\n",
      "\tat htsjdk.variant.variantcontext.Allele.<init>(Allele.java:162)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:61)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:70)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.insertAllele(Haplotype.java:236)\n",
      "\tat org.broadinstitute.gatk.utils.gga.GenotypingGivenAllelesUtils.composeGivenHaplotypes(GenotypingGivenAllelesUtils.java:132)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.LocalAssemblyEngine.runLocalAssembly(LocalAssemblyEngine.java:145)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.assembleReads(HaplotypeCaller.java:1029)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:865)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:228)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:709)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:705)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.executeSingleThreaded(NanoScheduler.java:274)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.execute(NanoScheduler.java:245)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:274)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:78)\n",
      "\tat org.broadinstitute.gatk.engine.executive.LinearMicroScheduler.execute(LinearMicroScheduler.java:99)\n",
      "\tat org.broadinstitute.gatk.engine.GenomeAnalysisEngine.execute(GenomeAnalysisEngine.java:315)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineExecutable.execute(CommandLineExecutable.java:121)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:248)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:155)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineGATK.main(CommandLineGATK.java:106)\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR A GATK RUNTIME ERROR has occurred (version 3.5-0-g36282e4):\n",
      "##### ERROR\n",
      "##### ERROR This might be a bug. Please check the documentation guide to see if this is a known problem.\n",
      "##### ERROR If not, please post the error message, with stack trace, to the GATK forum.\n",
      "##### ERROR Visit our website and forum for extensive documentation and answers to \n",
      "##### ERROR commonly asked questions http://www.broadinstitute.org/gatk\n",
      "##### ERROR\n",
      "##### ERROR MESSAGE: Unexpected base in allele bases 'GTGCACCTATGGTGCACTTTAAAAATAGTATCTGTAAACGAAGTGTTAGATAAACACGGCTTCGTGGTAATACCACATATATAGTGATACCACACATGTGATGTGACCCCACATTTACTGCATAGTGGTGCAACTTATTATATTGTGCATGGTGGTGCGACTTA*ATATTGGTGCGAATTTCTATTTTCGTGCATATTGGTGCGCTTTATAAAAAATTTTAGGATCTAGGTCTTGTCTCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTTTTTTAAATCTTAAAAATAGGAAAAACAAATGTACGTTTTAAAAAAAAGAAAAAAAAAAGAAAAAAAAAAAAAAATTTTTTTTTCTATGTCTTAAAAAGAGAAAAAGGAAATGTACATTTTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTTTTTGTCTCAAAAAAAAAAAATAGTG'\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam1_fn} -I {bam2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    -o {\"%s/test.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  12:34:19,818 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:34:19,821 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  12:34:19,821 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  12:34:19,822 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  12:34:19,822 HelpFormatter - [Fri Feb 03 12:34:19 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  12:34:19,822 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  12:34:19,827 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0388_C.bam -I /nfs/team112_internal/production_files/Pf/olivo_fanello_ringwald/1138_PF_CD_FANELLO/QG0327_C.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_01_v3.final.vcf.gz --intervals Pf3D7_01_v3:103000-104000 -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test2.vcf \n",
      "INFO  12:34:19,832 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  12:34:19,832 HelpFormatter - Date/Time: 2017/02/03 12:34:19 \n",
      "INFO  12:34:19,833 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:34:19,833 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:34:19,871 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:34:19,992 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  12:34:20,000 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  12:34:20,038 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  12:34:20,061 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "INFO  12:34:20,143 IntervalUtils - Processing 1001 bp from intervals \n",
      "WARN  12:34:20,148 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  12:34:20,263 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  12:34:20,294 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:34:20,295 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:34:20,295 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  12:34:20,296 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  12:34:20,297 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  12:34:20,365 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:34:20,371 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:34:20,434 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  12:34:28,693 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  12:34:28,694 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "WARN  12:34:32,584 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper \n",
      "##### ERROR --\n",
      "##### ERROR stack trace \n",
      "java.lang.IllegalArgumentException: Unexpected base in allele bases 'GTAAAATATATATATTAAGATGAGTACATGTTCATTTTATCTACATATTTATAATAAAGAAAATATTATTGCATTCTTTTCCTTATATATCTTTATACTTTGTTTTACACATATATTTTAGTTAAAATAAAATAAGGTAAGACAAATATTAGGAAGTCA*GTAAATGTAAAAAAATAAATCATATCTAAAATTTTATATAAATTTAAAAAAAAAAAAAAAAGAAAGAAAATATATAATACTTGACACATAAAATAAAATATTATATATATATATATAACATAATATTACTAATATTTATAAGTAATTATTCATATACAATAAAATGTTCATTTTGTTTCAAAAAAAAATATCTTTTACAGATAAAATATATAAATATATATATATATATATATATATATCTTTTTTTTTCAA'\n",
      "\tat htsjdk.variant.variantcontext.Allele.<init>(Allele.java:162)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:61)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:70)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.insertAllele(Haplotype.java:236)\n",
      "\tat org.broadinstitute.gatk.utils.gga.GenotypingGivenAllelesUtils.composeGivenHaplotypes(GenotypingGivenAllelesUtils.java:132)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.LocalAssemblyEngine.runLocalAssembly(LocalAssemblyEngine.java:145)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.assembleReads(HaplotypeCaller.java:1073)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:888)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:251)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:709)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:705)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.executeSingleThreaded(NanoScheduler.java:274)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.execute(NanoScheduler.java:245)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:274)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:78)\n",
      "\tat org.broadinstitute.gatk.engine.executive.LinearMicroScheduler.execute(LinearMicroScheduler.java:99)\n",
      "\tat org.broadinstitute.gatk.engine.GenomeAnalysisEngine.execute(GenomeAnalysisEngine.java:311)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineExecutable.execute(CommandLineExecutable.java:113)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:255)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:157)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineGATK.main(CommandLineGATK.java:108)\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR A GATK RUNTIME ERROR has occurred (version 3.6-0-g89b7209):\n",
      "##### ERROR\n",
      "##### ERROR This might be a bug. Please check the documentation guide to see if this is a known problem.\n",
      "##### ERROR If not, please post the error message, with stack trace, to the GATK forum.\n",
      "##### ERROR Visit our website and forum for extensive documentation and answers to \n",
      "##### ERROR commonly asked questions https://www.broadinstitute.org/gatk\n",
      "##### ERROR\n",
      "##### ERROR MESSAGE: Unexpected base in allele bases 'GTAAAATATATATATTAAGATGAGTACATGTTCATTTTATCTACATATTTATAATAAAGAAAATATTATTGCATTCTTTTCCTTATATATCTTTATACTTTGTTTTACACATATATTTTAGTTAAAATAAAATAAGGTAAGACAAATATTAGGAAGTCA*GTAAATGTAAAAAAATAAATCATATCTAAAATTTTATATAAATTTAAAAAAAAAAAAAAAAGAAAGAAAATATATAATACTTGACACATAAAATAAAATATTATATATATATATATAACATAATATTACTAATATTTATAAGTAATTATTCATATACAATAAAATGTTCATTTTGTTTCAAAAAAAAATATCTTTTACAGATAAAATATATAAATATATATATATATATATATATATATCTTTTTTTTTCAA'\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam1_fn} -I {bam2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    --intervals Pf3D7_01_v3:103000-104000 \\\n",
    "    -o {\"%s/test2.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  12:42:29,242 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:42:29,245 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  12:42:29,245 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  12:42:29,246 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  12:42:29,246 HelpFormatter - [Fri Feb 03 12:42:29 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  12:42:29,246 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  12:42:29,251 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam -I /lustre/scratch116/malaria/pfalciparum/output/4/2/a/9/36832/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_01_v3.final.vcf.gz --intervals Pf3D7_01_v3:100000-101000 -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test2.vcf \n",
      "INFO  12:42:29,256 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  12:42:29,257 HelpFormatter - Date/Time: 2017/02/03 12:42:29 \n",
      "INFO  12:42:29,258 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:42:29,258 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:42:29,295 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:42:29,421 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  12:42:29,429 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  12:42:29,533 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.10 \n",
      "INFO  12:42:29,615 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "INFO  12:42:29,690 IntervalUtils - Processing 1001 bp from intervals \n",
      "WARN  12:42:29,695 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  12:42:29,804 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  12:42:29,833 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:42:29,834 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:42:29,834 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  12:42:29,835 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  12:42:29,836 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  12:42:29,904 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:42:29,911 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:42:29,974 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "##### ERROR --\n",
      "##### ERROR stack trace \n",
      "java.lang.IllegalArgumentException: Unexpected base in allele bases 'TTCATGCAGAAAAATTATGAATAGTGACATTTCTTCCTTTAAACATATAAATGAATTGAAAAGTTTAGAACACAGAGCCGCAAAAGCTGCAGAAGCAGAAATGAAGAAAAGAGCTCAAAAACC*AAGAAGAAAAAAAGTAGAAGAGGATGGTTATGTTGTGGGGGGGGAGATATCGAAACAGTTGAACCACAACAAGAAGAACCAGTCCAAACCGTTCAAGAACAACAAGTAAATGAATATGGTGATATATTACCATCATTAAGGGCCAGTATTACTAATTCAGCTATTAATTATTATGATACCGTAAAAGATGGTGTATACTTAGACCATGAAACATCAGATGCTCTTTATACAGATGAAGATTTGTTATTTGATTTGGAAAAACAAAAATATATGGATATGTTAGATACATCTGAAGAAGAATCTGTTAAAGAAAATGAAGAAGAACACA'\n",
      "\tat htsjdk.variant.variantcontext.Allele.<init>(Allele.java:162)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:61)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:70)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.insertAllele(Haplotype.java:236)\n",
      "\tat org.broadinstitute.gatk.utils.gga.GenotypingGivenAllelesUtils.composeGivenHaplotypes(GenotypingGivenAllelesUtils.java:132)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.LocalAssemblyEngine.runLocalAssembly(LocalAssemblyEngine.java:145)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.assembleReads(HaplotypeCaller.java:1073)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:888)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:251)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:709)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:705)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.executeSingleThreaded(NanoScheduler.java:274)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.execute(NanoScheduler.java:245)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:274)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:78)\n",
      "\tat org.broadinstitute.gatk.engine.executive.LinearMicroScheduler.execute(LinearMicroScheduler.java:99)\n",
      "\tat org.broadinstitute.gatk.engine.GenomeAnalysisEngine.execute(GenomeAnalysisEngine.java:311)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineExecutable.execute(CommandLineExecutable.java:113)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:255)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:157)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineGATK.main(CommandLineGATK.java:108)\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR A GATK RUNTIME ERROR has occurred (version 3.6-0-g89b7209):\n",
      "##### ERROR\n",
      "##### ERROR This might be a bug. Please check the documentation guide to see if this is a known problem.\n",
      "##### ERROR If not, please post the error message, with stack trace, to the GATK forum.\n",
      "##### ERROR Visit our website and forum for extensive documentation and answers to \n",
      "##### ERROR commonly asked questions https://www.broadinstitute.org/gatk\n",
      "##### ERROR\n",
      "##### ERROR MESSAGE: Unexpected base in allele bases 'TTCATGCAGAAAAATTATGAATAGTGACATTTCTTCCTTTAAACATATAAATGAATTGAAAAGTTTAGAACACAGAGCCGCAAAAGCTGCAGAAGCAGAAATGAAGAAAAGAGCTCAAAAACC*AAGAAGAAAAAAAGTAGAAGAGGATGGTTATGTTGTGGGGGGGGAGATATCGAAACAGTTGAACCACAACAAGAAGAACCAGTCCAAACCGTTCAAGAACAACAAGTAAATGAATATGGTGATATATTACCATCATTAAGGGCCAGTATTACTAATTCAGCTATTAATTATTATGATACCGTAAAAGATGGTGTATACTTAGACCATGAAACATCAGATGCTCTTTATACAGATGAAGATTTGTTATTTGATTTGGAAAAACAAAAATATATGGATATGTTAGATACATCTGAAGAAGAATCTGTTAAAGAAAATGAAGAAGAACACA'\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} -I {bam_release_2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    --intervals Pf3D7_01_v3:100000-101000 \\\n",
    "    -o {\"%s/test2.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  12:44:38,627 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:44:38,629 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  12:44:38,630 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  12:44:38,630 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  12:44:38,631 HelpFormatter - [Fri Feb 03 12:44:38 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  12:44:38,631 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  12:44:38,636 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam -I /lustre/scratch116/malaria/pfalciparum/output/4/2/a/9/36832/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf_M76611.final.vcf.gz -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test_MT.vcf \n",
      "INFO  12:44:38,641 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  12:44:38,641 HelpFormatter - Date/Time: 2017/02/03 12:44:38 \n",
      "INFO  12:44:38,642 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:44:38,642 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  12:44:38,723 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:44:38,852 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  12:44:38,860 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  12:44:38,904 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  12:44:38,926 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "WARN  12:44:39,014 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  12:44:39,122 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  12:44:39,168 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:44:39,169 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:44:39,169 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  12:44:39,170 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  12:44:39,170 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  12:44:39,241 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:44:39,247 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  12:44:39,313 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "INFO  12:45:09,173 ProgressMeter - Pf3D7_01_v3:508912              0.0    30.0 s           49.6 w        2.2%    22.9 m      22.4 m \n",
      "INFO  12:45:39,175 ProgressMeter - Pf3D7_02_v3:404774         640851.0    60.0 s           93.0 s        4.5%    22.3 m      21.3 m \n",
      "INFO  12:46:09,177 ProgressMeter - Pf3D7_03_v3:17705        1587953.0    90.0 s           56.0 s        6.9%    21.8 m      20.3 m \n",
      "INFO  12:46:39,179 ProgressMeter - Pf3D7_03_v3:570906        1587953.0   120.0 s           75.0 s        9.3%    21.6 m      19.6 m \n",
      "INFO  12:47:09,180 ProgressMeter - Pf3D7_04_v3:29037        2655924.0     2.5 m           56.0 s       11.5%    21.7 m      19.2 m \n",
      "INFO  12:47:39,182 ProgressMeter - Pf3D7_04_v3:552756        2655924.0     3.0 m           67.0 s       13.8%    21.8 m      18.8 m \n",
      "INFO  12:48:09,183 ProgressMeter - Pf3D7_04_v3:1141908        2655924.0     3.5 m           79.0 s       16.3%    21.5 m      18.0 m \n",
      "INFO  12:48:39,185 ProgressMeter - Pf3D7_05_v3:489741        3856414.0     4.0 m           62.0 s       18.6%    21.5 m      17.5 m \n",
      "INFO  12:49:09,187 ProgressMeter - Pf3D7_05_v3:1018868        3856414.0     4.5 m           70.0 s       20.9%    21.5 m      17.0 m \n",
      "INFO  12:49:39,189 ProgressMeter - Pf3D7_06_v3:233442        5199971.0     5.0 m           57.0 s       23.3%    21.5 m      16.5 m \n",
      "INFO  12:50:09,190 ProgressMeter - Pf3D7_06_v3:783126        5199971.0     5.5 m           63.0 s       25.6%    21.4 m      15.9 m \n",
      "INFO  12:50:39,192 ProgressMeter - Pf3D7_06_v3:1335763        5199971.0     6.0 m           69.0 s       28.0%    21.4 m      15.4 m \n",
      "INFO  12:51:09,193 ProgressMeter - Pf3D7_07_v3:485120        6618213.0     6.5 m           58.0 s       30.4%    21.4 m      14.9 m \n",
      "INFO  12:51:39,195 ProgressMeter - Pf3D7_07_v3:1019000        6618213.0     7.0 m           63.0 s       32.7%    21.4 m      14.4 m \n",
      "INFO  12:52:09,201 ProgressMeter - Pf3D7_08_v3:108598        8063420.0     7.5 m           55.0 s       35.0%    21.4 m      13.9 m \n",
      "INFO  12:52:39,202 ProgressMeter - Pf3D7_08_v3:620688        8063420.0     8.0 m           59.0 s       37.2%    21.5 m      13.5 m \n",
      "INFO  12:53:09,204 ProgressMeter - Pf3D7_08_v3:1187566        8063420.0     8.5 m           63.0 s       39.6%    21.4 m      12.9 m \n",
      "INFO  12:53:39,205 ProgressMeter - Pf3D7_09_v3:261692        9536225.0     9.0 m           56.0 s       42.0%    21.4 m      12.4 m \n",
      "INFO  12:54:09,207 ProgressMeter - Pf3D7_09_v3:811180        9536225.0     9.5 m           59.0 s       44.3%    21.4 m      11.9 m \n",
      "INFO  12:54:39,209 ProgressMeter - Pf3D7_09_v3:1362796        9536225.0    10.0 m           62.0 s       46.7%    21.4 m      11.4 m \n",
      "INFO  12:55:09,210 ProgressMeter - Pf3D7_10_v3:396509       1.107796E7    10.5 m           56.0 s       49.2%    21.4 m      10.9 m \n",
      "INFO  12:55:39,212 ProgressMeter - Pf3D7_10_v3:938759       1.107796E7    11.0 m           59.0 s       51.5%    21.4 m      10.4 m \n",
      "INFO  12:56:09,213 ProgressMeter - Pf3D7_10_v3:1463943       1.107796E7    11.5 m           62.0 s       53.8%    21.4 m       9.9 m \n",
      "INFO  12:56:39,215 ProgressMeter - Pf3D7_11_v3:372826      1.2765616E7    12.0 m           56.0 s       56.3%    21.3 m       9.3 m \n",
      "INFO  12:57:09,216 ProgressMeter - Pf3D7_11_v3:912163      1.2765616E7    12.5 m           58.0 s       58.6%    21.3 m       8.8 m \n",
      "INFO  12:57:39,218 ProgressMeter - Pf3D7_11_v3:1457473      1.2765616E7    13.0 m           61.0 s       61.0%    21.3 m       8.3 m \n",
      "INFO  12:58:09,219 ProgressMeter - Pf3D7_11_v3:1985271      1.2765616E7    13.5 m           63.0 s       63.2%    21.4 m       7.9 m \n",
      "INFO  12:58:39,221 ProgressMeter - Pf3D7_12_v3:473160      1.4803956E7    14.0 m           56.0 s       65.5%    21.4 m       7.4 m \n",
      "INFO  12:59:09,222 ProgressMeter - Pf3D7_12_v3:1009878      1.4803956E7    14.5 m           58.0 s       67.8%    21.4 m       6.9 m \n",
      "INFO  12:59:39,224 ProgressMeter - Pf3D7_12_v3:1524153      1.4803956E7    15.0 m           60.0 s       70.0%    21.4 m       6.4 m \n",
      "INFO  13:00:09,225 ProgressMeter - Pf3D7_12_v3:2065052      1.4803956E7    15.5 m           62.0 s       72.3%    21.4 m       5.9 m \n",
      "INFO  13:00:39,227 ProgressMeter - Pf3D7_13_v3:380574       1.707545E7    16.0 m           56.0 s       74.8%    21.4 m       5.4 m \n",
      "INFO  13:01:09,228 ProgressMeter - Pf3D7_13_v3:908735       1.707545E7    16.5 m           57.0 s       77.1%    21.4 m       4.9 m \n",
      "INFO  13:01:39,230 ProgressMeter - Pf3D7_13_v3:1456674       1.707545E7    17.0 m           59.0 s       79.4%    21.4 m       4.4 m \n",
      "INFO  13:02:09,231 ProgressMeter - Pf3D7_13_v3:1992198       1.707545E7    17.5 m           61.0 s       81.7%    21.4 m       3.9 m \n",
      "INFO  13:02:39,238 ProgressMeter - Pf3D7_13_v3:2523900       1.707545E7    18.0 m           63.0 s       84.0%    21.4 m       3.4 m \n",
      "INFO  13:03:09,239 ProgressMeter - Pf3D7_14_v3:166846      2.0000686E7    18.5 m           55.0 s       86.4%    21.4 m       2.9 m \n",
      "INFO  13:03:39,240 ProgressMeter - Pf3D7_14_v3:701706      2.0000686E7    19.0 m           57.0 s       88.7%    21.4 m       2.4 m \n",
      "INFO  13:04:09,242 ProgressMeter - Pf3D7_14_v3:1234942      2.0000686E7    19.5 m           58.0 s       91.0%    21.4 m     115.0 s \n",
      "INFO  13:04:39,244 ProgressMeter - Pf3D7_14_v3:1754223      2.0000686E7    20.0 m           60.0 s       93.2%    21.5 m      87.0 s \n",
      "INFO  13:05:09,245 ProgressMeter - Pf3D7_14_v3:2280765      2.0000686E7    20.5 m           61.0 s       95.5%    21.5 m      58.0 s \n",
      "INFO  13:05:39,247 ProgressMeter - Pf3D7_14_v3:2814305      2.0000686E7    21.0 m           63.0 s       97.8%    21.5 m      28.0 s \n",
      "INFO  13:06:09,248 ProgressMeter -  Pf_M76611:1634      2.3326872E7    21.5 m           55.0 s      100.0%    21.5 m       0.0 s \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  13:06:17,302 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  13:06:17,302 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "INFO  13:06:17,303 VectorLoglessPairHMM - Time spent in setup for JNI call : 0.0 \n",
      "INFO  13:06:17,303 PairHMM - Total compute time in PairHMM computeLikelihoods() : 0.0 \n",
      "INFO  13:06:17,304 HaplotypeCaller - Ran local assembly on 0 active regions \n",
      "INFO  13:06:17,323 ProgressMeter -            done      2.3332839E7    21.6 m           55.0 s      100.0%    21.6 m       0.0 s \n",
      "INFO  13:06:17,324 ProgressMeter - Total runtime 1298.16 secs, 21.64 min, 0.36 hours \n",
      "INFO  13:06:17,324 MicroScheduler - 3602438 reads were filtered out during the traversal out of approximately 35392386 total reads (10.18%) \n",
      "INFO  13:06:17,325 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  13:06:17,325 MicroScheduler -   -> 314593 reads (0.89% of total) failing DuplicateReadFilter \n",
      "INFO  13:06:17,325 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  13:06:17,326 MicroScheduler -   -> 3210228 reads (9.07% of total) failing HCMappingQualityFilter \n",
      "INFO  13:06:17,326 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  13:06:17,326 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  13:06:17,326 MicroScheduler -   -> 77617 reads (0.22% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  13:06:17,327 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 1 WARN messages, the first 1 are repeated below.\n",
      "WARN  12:44:39,014 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} -I {bam_release_2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    -o {\"%s/test_MT.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  13:10:21,540 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:10:21,543 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  13:10:21,543 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  13:10:21,543 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  13:10:21,544 HelpFormatter - [Fri Feb 03 13:10:21 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  13:10:21,544 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  13:10:21,549 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam -I /lustre/scratch116/malaria/pfalciparum/output/4/2/a/9/36832/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_API_v3.final.vcf.gz -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test_API.vcf \n",
      "INFO  13:10:21,554 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  13:10:21,555 HelpFormatter - Date/Time: 2017/02/03 13:10:21 \n",
      "INFO  13:10:21,555 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:10:21,556 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:10:21,656 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  13:10:21,804 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  13:10:21,813 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:10:21,863 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.05 \n",
      "INFO  13:10:21,888 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "WARN  13:10:21,979 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  13:10:22,087 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  13:10:22,132 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  13:10:22,133 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  13:10:22,134 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  13:10:22,134 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  13:10:22,135 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  13:10:22,201 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  13:10:22,208 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  13:10:22,270 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "INFO  13:10:52,138 ProgressMeter - Pf3D7_01_v3:517534              0.0    30.0 s           49.6 w        2.2%    22.5 m      22.0 m \n",
      "INFO  13:11:22,140 ProgressMeter - Pf3D7_02_v3:422541         640851.0    60.0 s           93.0 s        4.6%    21.9 m      20.9 m \n",
      "INFO  13:11:52,143 ProgressMeter - Pf3D7_03_v3:33831        1587953.0    90.0 s           56.0 s        7.0%    21.6 m      20.1 m \n",
      "INFO  13:12:22,149 ProgressMeter - Pf3D7_03_v3:589843        1587953.0   120.0 s           75.0 s        9.3%    21.4 m      19.4 m \n",
      "INFO  13:12:52,151 ProgressMeter - Pf3D7_04_v3:39742        2655924.0     2.5 m           56.0 s       11.6%    21.6 m      19.1 m \n",
      "INFO  13:13:22,153 ProgressMeter - Pf3D7_04_v3:567193        2655924.0     3.0 m           67.0 s       13.8%    21.7 m      18.7 m \n",
      "INFO  13:13:52,155 ProgressMeter - Pf3D7_04_v3:1156132        2655924.0     3.5 m           79.0 s       16.3%    21.4 m      17.9 m \n",
      "INFO  13:14:22,157 ProgressMeter - Pf3D7_05_v3:492506        3856414.0     4.0 m           62.0 s       18.6%    21.5 m      17.5 m \n",
      "INFO  13:14:52,158 ProgressMeter - Pf3D7_05_v3:1026903        3856414.0     4.5 m           70.0 s       20.9%    21.5 m      17.0 m \n",
      "INFO  13:15:22,160 ProgressMeter - Pf3D7_06_v3:237034        5199971.0     5.0 m           57.0 s       23.3%    21.5 m      16.5 m \n",
      "INFO  13:15:52,162 ProgressMeter - Pf3D7_06_v3:790243        5199971.0     5.5 m           63.0 s       25.7%    21.4 m      15.9 m \n",
      "INFO  13:16:22,164 ProgressMeter - Pf3D7_06_v3:1336637        5199971.0     6.0 m           69.0 s       28.0%    21.4 m      15.4 m \n",
      "INFO  13:16:52,165 ProgressMeter - Pf3D7_07_v3:485514        6618213.0     6.5 m           58.0 s       30.4%    21.3 m      14.8 m \n",
      "INFO  13:17:22,167 ProgressMeter - Pf3D7_07_v3:1017677        6618213.0     7.0 m           63.0 s       32.7%    21.4 m      14.4 m \n",
      "INFO  13:17:52,169 ProgressMeter - Pf3D7_08_v3:101343        8063420.0     7.5 m           55.0 s       35.0%    21.4 m      13.9 m \n",
      "INFO  13:18:22,170 ProgressMeter - Pf3D7_08_v3:611965        8063420.0     8.0 m           59.0 s       37.2%    21.5 m      13.5 m \n",
      "INFO  13:18:52,172 ProgressMeter - Pf3D7_08_v3:1173227        8063420.0     8.5 m           63.0 s       39.6%    21.5 m      13.0 m \n",
      "INFO  13:19:22,174 ProgressMeter - Pf3D7_09_v3:255906        9536225.0     9.0 m           56.0 s       42.0%    21.4 m      12.4 m \n",
      "INFO  13:19:52,175 ProgressMeter - Pf3D7_09_v3:809559        9536225.0     9.5 m           59.0 s       44.3%    21.4 m      11.9 m \n",
      "INFO  13:20:22,177 ProgressMeter - Pf3D7_09_v3:1357643        9536225.0    10.0 m           62.0 s       46.7%    21.4 m      11.4 m \n",
      "INFO  13:20:52,179 ProgressMeter - Pf3D7_10_v3:395696       1.107796E7    10.5 m           56.0 s       49.2%    21.4 m      10.9 m \n",
      "INFO  13:21:22,181 ProgressMeter - Pf3D7_10_v3:933575       1.107796E7    11.0 m           59.0 s       51.5%    21.4 m      10.4 m \n",
      "INFO  13:21:52,182 ProgressMeter - Pf3D7_10_v3:1463683       1.107796E7    11.5 m           62.0 s       53.8%    21.4 m       9.9 m \n",
      "INFO  13:22:22,184 ProgressMeter - Pf3D7_11_v3:378671      1.2765616E7    12.0 m           56.0 s       56.3%    21.3 m       9.3 m \n",
      "INFO  13:22:52,185 ProgressMeter - Pf3D7_11_v3:918886      1.2765616E7    12.5 m           58.0 s       58.6%    21.3 m       8.8 m \n",
      "INFO  13:23:22,187 ProgressMeter - Pf3D7_11_v3:1471109      1.2765616E7    13.0 m           61.0 s       61.0%    21.3 m       8.3 m \n",
      "INFO  13:23:52,189 ProgressMeter - Pf3D7_11_v3:2004918      1.2765616E7    13.5 m           63.0 s       63.3%    21.3 m       7.8 m \n",
      "INFO  13:24:22,190 ProgressMeter - Pf3D7_12_v3:502028      1.4803956E7    14.0 m           56.0 s       65.6%    21.3 m       7.3 m \n",
      "INFO  13:24:52,192 ProgressMeter - Pf3D7_12_v3:1051542      1.4803956E7    14.5 m           58.0 s       68.0%    21.3 m       6.8 m \n",
      "INFO  13:25:22,193 ProgressMeter - Pf3D7_12_v3:1571719      1.4803956E7    15.0 m           60.0 s       70.2%    21.4 m       6.4 m \n",
      "INFO  13:25:52,195 ProgressMeter - Pf3D7_12_v3:2117567      1.4803956E7    15.5 m           62.0 s       72.5%    21.4 m       5.9 m \n",
      "INFO  13:26:22,196 ProgressMeter - Pf3D7_13_v3:442056       1.707545E7    16.0 m           56.0 s       75.1%    21.3 m       5.3 m \n",
      "INFO  13:26:52,198 ProgressMeter - Pf3D7_13_v3:984844       1.707545E7    16.5 m           57.0 s       77.4%    21.3 m       4.8 m \n",
      "INFO  13:27:22,199 ProgressMeter - Pf3D7_13_v3:1532715       1.707545E7    17.0 m           59.0 s       79.8%    21.3 m       4.3 m \n",
      "INFO  13:27:52,200 ProgressMeter - Pf3D7_13_v3:2069357       1.707545E7    17.5 m           61.0 s       82.1%    21.3 m       3.8 m \n",
      "INFO  13:28:22,202 ProgressMeter - Pf3D7_13_v3:2592012       1.707545E7    18.0 m           63.0 s       84.3%    21.4 m       3.4 m \n",
      "INFO  13:28:52,203 ProgressMeter - Pf3D7_14_v3:250365      2.0000686E7    18.5 m           55.0 s       86.8%    21.3 m       2.8 m \n",
      "INFO  13:29:22,204 ProgressMeter - Pf3D7_14_v3:795217      2.0000686E7    19.0 m           57.0 s       89.1%    21.3 m       2.3 m \n",
      "INFO  13:29:52,206 ProgressMeter - Pf3D7_14_v3:1336164      2.0000686E7    19.5 m           58.0 s       91.4%    21.3 m     109.0 s \n",
      "INFO  13:30:22,207 ProgressMeter - Pf3D7_14_v3:1857490      2.0000686E7    20.0 m           60.0 s       93.7%    21.3 m      80.0 s \n",
      "INFO  13:30:52,209 ProgressMeter - Pf3D7_14_v3:2388034      2.0000686E7    20.5 m           61.0 s       96.0%    21.4 m      51.0 s \n",
      "INFO  13:31:22,211 ProgressMeter - Pf3D7_14_v3:2927255      2.0000686E7    21.0 m           63.0 s       98.3%    21.4 m      22.0 s \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  13:31:42,864 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  13:31:42,865 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "WARN  13:31:43,414 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper \n",
      "##### ERROR --\n",
      "##### ERROR stack trace \n",
      "java.lang.IllegalArgumentException: Unexpected base in allele bases 'ATGATAAAATTTTTAAAACCTAAAATAAAAATATTAAAAAAATTAAATATACCTTTTTTATTATATTTATCTAGTAAATATAATTATAAATATTTAAATTATAAAATTTCATATAAATCTTATTTTGATTTAAAATTAAAATTTATTAGATATATATGTTATAATTATTGTATAACATATAAAAAATATTTATATTATTTGAATAAAATA*ATAATAAAAATATAAATATTTTATATTTTAAATTATTAAAAATATTAGAATTAAGATTGGATATATTTTTAGTTAATATAGGTTTTTTTAAAACTATATTGCAATCAAGGTATTATATTAAATATAAAAATATTTATATTAATAATATTATAAATAAATATTATAATATTAATTTAAAAAATAATGATATTTTATTTTTTAATAATAAAATAAAATATATAATATTAAAAAATTTAATT'\n",
      "\tat htsjdk.variant.variantcontext.Allele.<init>(Allele.java:162)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:61)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:70)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.insertAllele(Haplotype.java:236)\n",
      "\tat org.broadinstitute.gatk.utils.gga.GenotypingGivenAllelesUtils.composeGivenHaplotypes(GenotypingGivenAllelesUtils.java:132)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.LocalAssemblyEngine.runLocalAssembly(LocalAssemblyEngine.java:145)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.assembleReads(HaplotypeCaller.java:1073)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:888)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:251)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:709)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:705)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.executeSingleThreaded(NanoScheduler.java:274)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.execute(NanoScheduler.java:245)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:274)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:78)\n",
      "\tat org.broadinstitute.gatk.engine.executive.LinearMicroScheduler.execute(LinearMicroScheduler.java:99)\n",
      "\tat org.broadinstitute.gatk.engine.GenomeAnalysisEngine.execute(GenomeAnalysisEngine.java:311)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineExecutable.execute(CommandLineExecutable.java:113)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:255)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:157)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineGATK.main(CommandLineGATK.java:108)\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR A GATK RUNTIME ERROR has occurred (version 3.6-0-g89b7209):\n",
      "##### ERROR\n",
      "##### ERROR This might be a bug. Please check the documentation guide to see if this is a known problem.\n",
      "##### ERROR If not, please post the error message, with stack trace, to the GATK forum.\n",
      "##### ERROR Visit our website and forum for extensive documentation and answers to \n",
      "##### ERROR commonly asked questions https://www.broadinstitute.org/gatk\n",
      "##### ERROR\n",
      "##### ERROR MESSAGE: Unexpected base in allele bases 'ATGATAAAATTTTTAAAACCTAAAATAAAAATATTAAAAAAATTAAATATACCTTTTTTATTATATTTATCTAGTAAATATAATTATAAATATTTAAATTATAAAATTTCATATAAATCTTATTTTGATTTAAAATTAAAATTTATTAGATATATATGTTATAATTATTGTATAACATATAAAAAATATTTATATTATTTGAATAAAATA*ATAATAAAAATATAAATATTTTATATTTTAAATTATTAAAAATATTAGAATTAAGATTGGATATATTTTTAGTTAATATAGGTTTTTTTAAAACTATATTGCAATCAAGGTATTATATTAAATATAAAAATATTTATATTAATAATATTATAAATAAATATTATAATATTAATTTAAAAAATAATGATATTTTATTTTTTAATAATAAAATAAAATATATAATATTAAAAAATTTAATT'\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} -I {bam_release_2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    -o {\"%s/test_API.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  13:31:47,113 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:31:47,116 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  13:31:47,116 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  13:31:47,117 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  13:31:47,117 HelpFormatter - [Fri Feb 03 13:31:47 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  13:31:47,118 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  13:31:47,123 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam -I /lustre/scratch116/malaria/pfalciparum/output/4/2/a/9/36832/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_API_v3.final.vcf.gz --intervals Pf3D7_01_v3:200000-201000 -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test_200000.vcf \n",
      "INFO  13:31:47,129 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  13:31:47,129 HelpFormatter - Date/Time: 2017/02/03 13:31:47 \n",
      "INFO  13:31:47,130 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:31:47,130 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  13:31:47,170 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  13:31:47,300 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  13:31:47,308 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:31:47,347 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  13:31:47,368 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "INFO  13:31:47,438 IntervalUtils - Processing 1001 bp from intervals \n",
      "WARN  13:31:47,442 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  13:31:47,548 GenomeAnalysisEngine - Preparing for traversal over 2 BAM files \n",
      "INFO  13:31:47,575 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  13:31:47,576 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  13:31:47,576 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  13:31:47,577 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  13:31:47,578 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  13:31:47,652 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  13:31:47,659 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  13:31:47,730 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  13:31:48,839 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  13:31:48,840 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "INFO  13:31:48,841 VectorLoglessPairHMM - Time spent in setup for JNI call : 0.0 \n",
      "INFO  13:31:48,841 PairHMM - Total compute time in PairHMM computeLikelihoods() : 0.0 \n",
      "INFO  13:31:48,842 HaplotypeCaller - Ran local assembly on 0 active regions \n",
      "INFO  13:31:48,868 ProgressMeter -            done           1001.0     1.0 s           21.5 m       99.9%     1.0 s       0.0 s \n",
      "INFO  13:31:48,869 ProgressMeter - Total runtime 1.29 secs, 0.02 min, 0.00 hours \n",
      "INFO  13:31:48,870 MicroScheduler - 31 reads were filtered out during the traversal out of approximately 2324 total reads (1.33%) \n",
      "INFO  13:31:48,871 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  13:31:48,871 MicroScheduler -   -> 16 reads (0.69% of total) failing DuplicateReadFilter \n",
      "INFO  13:31:48,871 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  13:31:48,872 MicroScheduler -   -> 13 reads (0.56% of total) failing HCMappingQualityFilter \n",
      "INFO  13:31:48,873 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  13:31:48,873 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  13:31:48,874 MicroScheduler -   -> 2 reads (0.09% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  13:31:48,874 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 1 WARN messages, the first 1 are repeated below.\n",
      "WARN  13:31:47,442 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} -I {bam_release_2_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    --intervals Pf3D7_01_v3:200000-201000 \\\n",
    "    -o {\"%s/test_200000.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  14:55:55,063 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:55:55,066 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  14:55:55,067 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  14:55:55,067 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  14:55:55,067 HelpFormatter - [Mon Feb 06 14:55:55 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  14:55:55,068 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  14:55:55,073 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_API_v3.final.vcf.gz --intervals Pf3D7_01_v3:200000-201000 -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test_1sample_200000.vcf \n",
      "INFO  14:55:55,079 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  14:55:55,079 HelpFormatter - Date/Time: 2017/02/06 14:55:55 \n",
      "INFO  14:55:55,080 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:55:55,080 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:55:55,139 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  14:55:55,350 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  14:55:55,358 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  14:55:55,404 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.04 \n",
      "INFO  14:55:55,435 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "INFO  14:55:55,510 IntervalUtils - Processing 1001 bp from intervals \n",
      "WARN  14:55:55,515 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  14:55:55,725 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  14:55:55,745 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  14:55:55,746 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  14:55:55,746 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  14:55:55,747 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  14:55:55,747 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  14:55:55,816 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  14:55:55,822 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  14:55:55,887 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  14:55:56,443 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  14:55:56,443 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "INFO  14:55:56,444 VectorLoglessPairHMM - Time spent in setup for JNI call : 0.0 \n",
      "INFO  14:55:56,445 PairHMM - Total compute time in PairHMM computeLikelihoods() : 0.0 \n",
      "INFO  14:55:56,445 HaplotypeCaller - Ran local assembly on 0 active regions \n",
      "INFO  14:55:56,464 ProgressMeter -            done           1001.0     0.0 s           11.9 m       99.9%     0.0 s       0.0 s \n",
      "INFO  14:55:56,465 ProgressMeter - Total runtime 0.72 secs, 0.01 min, 0.00 hours \n",
      "INFO  14:55:56,466 MicroScheduler - 9 reads were filtered out during the traversal out of approximately 470 total reads (1.91%) \n",
      "INFO  14:55:56,466 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  14:55:56,467 MicroScheduler -   -> 4 reads (0.85% of total) failing DuplicateReadFilter \n",
      "INFO  14:55:56,467 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  14:55:56,468 MicroScheduler -   -> 5 reads (1.06% of total) failing HCMappingQualityFilter \n",
      "INFO  14:55:56,468 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  14:55:56,469 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  14:55:56,469 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  14:55:56,469 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 1 WARN messages, the first 1 are repeated below.\n",
      "WARN  14:55:55,515 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    --intervals Pf3D7_01_v3:200000-201000 \\\n",
    "    -o {\"%s/test_1sample_200000.vcf\" % temp_dir}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  14:59:28,737 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:59:28,740 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.6-0-g89b7209, Compiled 2016/06/01 22:27:29 \n",
      "INFO  14:59:28,740 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  14:59:28,740 HelpFormatter - For support and documentation go to https://www.broadinstitute.org/gatk \n",
      "INFO  14:59:28,741 HelpFormatter - [Mon Feb 06 14:59:28 GMT 2017] Executing on Linux 3.2.0-105-generic amd64 \n",
      "INFO  14:59:28,741 HelpFormatter - Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02 JdkDeflater \n",
      "INFO  14:59:28,746 HelpFormatter - Program Args: -T HaplotypeCaller -R /lustre/scratch116/malaria/pfalciparum/resources/Pfalciparum.genome.fasta -I /lustre/scratch116/malaria/pfalciparum/output/4/8/8/2/36831/4_bam_mark_duplicates_v2/pe.1.markdup.bam --genotyping_mode GENOTYPE_GIVEN_ALLELES --output_mode EMIT_ALL_SITES --alleles /nfs/team112_internal/production/release_build/Pf/6_0_release_packages/vcf/Pf_60_Pf3D7_API_v3.final.vcf.gz -o /lustre/scratch116/malaria/rp7/data/methods-dev/20170203_genotype_at_known_loci_test/test_1sample_API.vcf \n",
      "INFO  14:59:28,751 HelpFormatter - Executing as rp7@bc-31-1-16 on Linux 3.2.0-105-generic amd64; Java HotSpot(TM) 64-Bit Server VM 1.8.0_74-b02. \n",
      "INFO  14:59:28,752 HelpFormatter - Date/Time: 2017/02/06 14:59:28 \n",
      "INFO  14:59:28,752 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:59:28,753 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  14:59:28,788 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  14:59:28,914 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  14:59:28,922 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  14:59:28,951 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.03 \n",
      "INFO  14:59:28,967 HCMappingQualityFilter - Filtering out reads with MAPQ < 20 \n",
      "WARN  14:59:29,054 IndexDictionaryUtils - Track alleles doesn't have a sequence dictionary built in, skipping dictionary validation \n",
      "INFO  14:59:29,160 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  14:59:29,192 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  14:59:29,193 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  14:59:29,193 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  14:59:29,194 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  14:59:29,195 HaplotypeCaller - Disabling physical phasing, which is supported only for reference-model confidence output \n",
      "INFO  14:59:29,264 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  14:59:29,271 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  14:59:29,336 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "INFO  14:59:59,198 ProgressMeter - Pf3D7_03_v3:913542        1587953.0    30.0 s           18.0 s       10.7%     4.7 m       4.2 m \n",
      "INFO  15:00:29,200 ProgressMeter - Pf3D7_05_v3:1262980        3856414.0    60.0 s           15.0 s       21.9%     4.6 m       3.6 m \n",
      "INFO  15:00:59,201 ProgressMeter - Pf3D7_07_v3:1027031        6618213.0    90.0 s           13.0 s       32.8%     4.6 m       3.1 m \n",
      "INFO  15:01:29,203 ProgressMeter - Pf3D7_09_v3:736212        9536225.0   120.0 s           12.0 s       44.0%     4.5 m       2.5 m \n",
      "INFO  15:01:59,205 ProgressMeter - Pf3D7_11_v3:258977      1.2765616E7     2.5 m           11.0 s       55.8%     4.5 m     118.0 s \n",
      "INFO  15:02:29,207 ProgressMeter - Pf3D7_12_v3:937800      1.4803956E7     3.0 m           12.0 s       67.5%     4.4 m      86.0 s \n",
      "INFO  15:02:59,208 ProgressMeter - Pf3D7_13_v3:1329935       1.707545E7     3.5 m           12.0 s       78.9%     4.4 m      56.0 s \n",
      "INFO  15:03:29,210 ProgressMeter - Pf3D7_14_v3:1167540      2.0000686E7     4.0 m           12.0 s       90.7%     4.4 m      24.0 s \n",
      "Using AVX accelerated implementation of PairHMM\n",
      "INFO  15:03:54,614 VectorLoglessPairHMM - libVectorLoglessPairHMM unpacked successfully from GATK jar file \n",
      "INFO  15:03:54,614 VectorLoglessPairHMM - Using vectorized implementation of PairHMM \n",
      "WARN  15:03:54,931 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper \n",
      "##### ERROR --\n",
      "##### ERROR stack trace \n",
      "java.lang.IllegalArgumentException: Unexpected base in allele bases 'ATGATAAAATTTTTAAAACCTAAAATAAAAATATTAAAAAAATTAAATATACCTTTTTTATTATATTTATCTAGTAAATATAATTATAAATATTTAAATTATAAAATTTCATATAAATCTTATTTTGATTTAAAATTAAAATTTATTAGATATATATGTTATAATTATTGTATAACATATAAAAAATATTTATATTATTTGAATAAAATA*ATAATAAAAATATAAATATTTTATATTTTAAATTATTAAAAATATTAGAATTAAGATTGGATATATTTTTAGTTAATATAGGTTTTTTTAAAACTATATTGCAATCAAGGTATTATATTAAATATAAAAATATTTATATTAATAATATTATAAATAAATATTATAATATTAATTTAAAAAATAATGATATTTTATTTTTTAATAATAAAATAAAATATATAATATTAAAAAATTTAATT'\n",
      "\tat htsjdk.variant.variantcontext.Allele.<init>(Allele.java:162)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:61)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.<init>(Haplotype.java:70)\n",
      "\tat org.broadinstitute.gatk.utils.haplotype.Haplotype.insertAllele(Haplotype.java:236)\n",
      "\tat org.broadinstitute.gatk.utils.gga.GenotypingGivenAllelesUtils.composeGivenHaplotypes(GenotypingGivenAllelesUtils.java:132)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.LocalAssemblyEngine.runLocalAssembly(LocalAssemblyEngine.java:145)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.assembleReads(HaplotypeCaller.java:1073)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:888)\n",
      "\tat org.broadinstitute.gatk.tools.walkers.haplotypecaller.HaplotypeCaller.map(HaplotypeCaller.java:251)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:709)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions$TraverseActiveRegionMap.apply(TraverseActiveRegions.java:705)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.executeSingleThreaded(NanoScheduler.java:274)\n",
      "\tat org.broadinstitute.gatk.utils.nanoScheduler.NanoScheduler.execute(NanoScheduler.java:245)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:274)\n",
      "\tat org.broadinstitute.gatk.engine.traversals.TraverseActiveRegions.traverse(TraverseActiveRegions.java:78)\n",
      "\tat org.broadinstitute.gatk.engine.executive.LinearMicroScheduler.execute(LinearMicroScheduler.java:99)\n",
      "\tat org.broadinstitute.gatk.engine.GenomeAnalysisEngine.execute(GenomeAnalysisEngine.java:311)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineExecutable.execute(CommandLineExecutable.java:113)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:255)\n",
      "\tat org.broadinstitute.gatk.utils.commandline.CommandLineProgram.start(CommandLineProgram.java:157)\n",
      "\tat org.broadinstitute.gatk.engine.CommandLineGATK.main(CommandLineGATK.java:108)\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n",
      "##### ERROR A GATK RUNTIME ERROR has occurred (version 3.6-0-g89b7209):\n",
      "##### ERROR\n",
      "##### ERROR This might be a bug. Please check the documentation guide to see if this is a known problem.\n",
      "##### ERROR If not, please post the error message, with stack trace, to the GATK forum.\n",
      "##### ERROR Visit our website and forum for extensive documentation and answers to \n",
      "##### ERROR commonly asked questions https://www.broadinstitute.org/gatk\n",
      "##### ERROR\n",
      "##### ERROR MESSAGE: Unexpected base in allele bases 'ATGATAAAATTTTTAAAACCTAAAATAAAAATATTAAAAAAATTAAATATACCTTTTTTATTATATTTATCTAGTAAATATAATTATAAATATTTAAATTATAAAATTTCATATAAATCTTATTTTGATTTAAAATTAAAATTTATTAGATATATATGTTATAATTATTGTATAACATATAAAAAATATTTATATTATTTGAATAAAATA*ATAATAAAAATATAAATATTTTATATTTTAAATTATTAAAAATATTAGAATTAAGATTGGATATATTTTTAGTTAATATAGGTTTTTTTAAAACTATATTGCAATCAAGGTATTATATTAAATATAAAAATATTTATATTAATAATATTATAAATAAATATTATAATATTAATTTAAAAAATAATGATATTTTATTTTTTAATAATAAAATAAAATATATAATATTAAAAAATTTAATT'\n",
      "##### ERROR ------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!{GenomeAnalysisTK} \\\n",
    "     -T HaplotypeCaller \\\n",
    "     -R {genome_fn} \\\n",
    "     -I {bam_release_1_fn} \\\n",
    "    --genotyping_mode GENOTYPE_GIVEN_ALLELES \\\n",
    "    --output_mode EMIT_ALL_SITES \\\n",
    "    --alleles {release_vcf_fn} \\\n",
    "    -o {\"%s/test_1sample_API.vcf\" % temp_dir}\n",
    "#     --intervals Pf3D7_01_v3:200000-201000 \\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
