{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Plan\n",
    "- Figure out how /nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/SNP_INDEL_WG.combined.filtered.crosses.vcf.gz was created and document here\n",
    "- Filter out AC=0 to get file size down and remove unnecessary INFO fields\n",
    "- Annotate each variant with a multiallelic flag by filter multiallelic to sep vcf then bcftools annotate with this\n",
    "- Create biallelic (and left aligned?) version using bcftools norm\n",
    "- Annotate STRs with GATK\n",
    "- Create SNP/indel/STR by filtering out different sets and setting appropriate INFO with bcftools annotate\n",
    "- Annotate coding flag by filtering to new vcf and then bcftools annotate\n",
    "\n",
    "\n",
    "- Create function to create biallelic, 5/2 rule, new AF, segregating, minimal, renormalised VCF\n",
    "- Split the above into SNPs and INDELs\n",
    "- Test function on small subset of chr14\n",
    "- Run function on chrom 14\n",
    "- New function to also create npy file\n",
    "- Read in chr14 npy file, and calculate Mendelian error and genotype concordance\n",
    "- Attempt to reannotate above with STR and SNPEFF annotations\n",
    "- Rerun scripts to get breakdown by SNP/STR/nonSTR coding/noncoding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See 20160203_release5_npy_hdf5.ipynb for creation of VCF specific to crosses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python 3.4.3 |Anaconda 2.2.0 (64-bit)| (default, Mar  6 2015, 12:03:53) \n",
      "[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]\n",
      "numpy 1.9.2\n",
      "scipy 0.15.1\n",
      "pandas 0.15.2\n",
      "numexpr 2.3.1\n",
      "pysam 0.8.3\n",
      "pysamstats 0.23\n",
      "petl 1.0.11\n",
      "petlx 1.0.3\n",
      "vcf 0.6.7\n",
      "vcfnp 2.3.0.dev0\n",
      "h5py 2.4.0\n",
      "tables 3.1.1\n"
     ]
    }
   ],
   "source": [
    "%run _standard_imports.ipynb\n",
    "%run _shared_setup.ipynb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "release5_final_files_dir = '/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0'\n",
    "# chrom_vcf_fn = \"%s/SNP_INDEL_Pf3D7_14_v3.combined.filtered.vcf.gz\" % (release5_final_files_dir)\n",
    "crosses_vcf_fn = \"%s/SNP_INDEL_WG.combined.filtered.crosses.vcf.gz\" % (release5_final_files_dir)\n",
    "sites_only_vcf_fn = \"%s/SNP_INDEL_WG.combined.filtered.sites.vcf.gz\" % (release5_final_files_dir)\n",
    "\n",
    "output_dir = \"/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance\"\n",
    "!mkdir -p {output_dir}/vcf\n",
    "\n",
    "release5_crosses_metadata_txt_fn = '../../meta/pf3k_release_5_crosses_metadata.txt'\n",
    "gff_fn = \"%s/Pfalciparum.noseq.gff3.gz\" % output_dir\n",
    "cds_gff_fn = \"%s/Pfalciparum.noseq.gff3.cds.gz\" % output_dir\n",
    "\n",
    "results_table_fn = \"%s/genotype_quality.xlsx\" % output_dir\n",
    "counts_table_fn = \"%s/variant_counts.xlsx\" % output_dir\n",
    "\n",
    "BCFTOOLS = '/nfs/team112_internal/rp7/src/github/malariagen/methods-dev/pf3k_techbm/opt_4/bcftools/bcftools'\n",
    "GATK = '/software/jre1.7.0_25/bin/java -jar /nfs/team112_internal/production/tools/bin/gatk/GenomeAnalysisTK-3.4-46/GenomeAnalysisTK.jar'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/Pfalciparum.noseq.gff3.gz'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# !wget ftp://ftp.sanger.ac.uk/pub/project/pathogens/gff3/2016-06/Pfalciparum.noseq.gff3.gz \\\n",
    "#     -O {gff_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# !zgrep CDS {gff_fn} | sort -k1,1 -k4n,5n | cut -f 1,4,5 | sed 's/$/\\t1/' | bgzip -c > {cds_gff_fn} && tabix -s1 -b2 -e3 {cds_gff_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/SNP_INDEL_WG.combined.filtered.crosses.vcf.gz'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crosses_vcf_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "multiallelic_header_fn = \"%s/vcf/MULTIALLELIC.hdr\" % (output_dir)\n",
    "fo=open(multiallelic_header_fn, 'w')\n",
    "print('##INFO=<ID=MULTIALLELIC,Number=0,Type=Flag,Description=\"Is position multiallelic\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "variant_type_header_fn = \"%s/vcf/VARIANT_TYPE.hdr\" % (output_dir)\n",
    "fo=open(variant_type_header_fn, 'w')\n",
    "print('##INFO=<ID=VARIANT_TYPE,Number=1,Type=String,Description=\"SNP, INDEL or STR\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cds_header_fn = \"%s/vcf/CDS.hdr\" % (output_dir)\n",
    "fo=open(cds_header_fn, 'w')\n",
    "print('##INFO=<ID=CDS,Number=0,Type=Flag,Description=\"Is position coding\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_analysis_vcf(input_vcf_fn=crosses_vcf_fn, region='Pf3D7_14_v3:1000000-1100000',\n",
    "                        output_vcf_fn=None, BCFTOOLS=BCFTOOLS, rewrite=False):\n",
    "    if output_vcf_fn is None:\n",
    "        if region is None:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir)\n",
    "        else:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_%s.analysis.vcf.gz\" % (output_dir, region)\n",
    "    intermediate_fns = collections.OrderedDict()\n",
    "    for intermediate_file in ['nonref', 'multiallelic', 'multiallelics', 'biallelic', 'str', 'snps', 'indels', 'strs',\n",
    "                              'variant_type', 'coding', 'analysis']:\n",
    "        intermediate_fns[intermediate_file] = output_vcf_fn.replace('analysis', intermediate_file)\n",
    "    \n",
    "    if rewrite or not os.path.exists(intermediate_fns['nonref']):\n",
    "        if region is not None:\n",
    "            !{BCFTOOLS} annotate --regions {region} --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set,\\\n",
    "FORMAT/PGT,FORMAT/PID,FORMAT/PL {input_vcf_fn} | \\\n",
    "            {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['nonref']}\n",
    "        else:\n",
    "            !{BCFTOOLS} annotate --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set,\\\n",
    "FORMAT/PGT,FORMAT/PID,FORMAT/PL {input_vcf_fn} | \\\n",
    "            {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['nonref']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['nonref']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t1\\n' --include 'N_ALT>1' {intermediate_fns['nonref']} | bgzip -c > {intermediate_fns['multiallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['multiallelic']}\n",
    "#         !{BCFTOOLS} index --tbi {intermediate_fns['multiallelic']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelics']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['multiallelic']} -c CHROM,POS,INFO/MULTIALLELIC \\\n",
    "        -h {multiallelic_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['multiallelics']} {intermediate_fns['nonref']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['multiallelics']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['biallelic']):\n",
    "        !{BCFTOOLS} norm -m -any --fasta-ref {GENOME_FN} {intermediate_fns['multiallelics']} | \\\n",
    "        {BCFTOOLS} view --include 'AC>0 && ALT!=\"*\"' -Oz -o {intermediate_fns['biallelic']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['biallelic']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['str']):\n",
    "        !{GATK} -T VariantAnnotator \\\n",
    "            -R {GENOME_FN} \\\n",
    "            -o {intermediate_fns['str']} \\\n",
    "            -A TandemRepeatAnnotator  \\\n",
    "            -V {intermediate_fns['biallelic']}\n",
    "#         !{BCFTOOLS} index --tbi {intermediate_fns['str']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['snps']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSNP\\n' --include 'TYPE=\"snp\"' {intermediate_fns['str']} | bgzip -c > {intermediate_fns['snps']} && tabix -s1 -b2 -e2 -f {intermediate_fns['snps']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['indels']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tINDEL\\n' --include 'TYPE!=\"snp\" && STR=0' {intermediate_fns['str']} | bgzip -c > {intermediate_fns['indels']} && tabix -s1 -b2 -e2 -f {intermediate_fns['indels']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['strs']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSTR\\n' --include 'TYPE!=\"snp\" && STR=1' {intermediate_fns['str']} | bgzip -c > {intermediate_fns['strs']} && tabix -s1 -b2 -e2 -f {intermediate_fns['strs']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['variant_type']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['snps']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -h {variant_type_header_fn} {intermediate_fns['str']} |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['indels']} -c CHROM,POS,INFO/VARIANT_TYPE |\\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['strs']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -Oz -o {intermediate_fns['variant_type']} \n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['variant_type']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['analysis']):\n",
    "        !{BCFTOOLS} annotate -a {cds_gff_fn} -c CHROM,FROM,TO,CDS \\\n",
    "        -h {cds_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['analysis']} {intermediate_fns['variant_type']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['analysis']}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "create_analysis_vcf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "create_analysis_vcf(region=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "98\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>sample</th>\n",
       "<th>acc</th>\n",
       "<th>study</th>\n",
       "<th>study_title</th>\n",
       "<th>clone</th>\n",
       "<th>parent_or_progeny</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>PG0008-CW</td>\n",
       "<td>ERS010540</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>DD2</td>\n",
       "<td>parent</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0004-CW</td>\n",
       "<td>ERS010539</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>HB3</td>\n",
       "<td>parent</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0023-C</td>\n",
       "<td>ERS010581,ERS010543</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>1BB5</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0022-Cx</td>\n",
       "<td>ERS088714</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>3BA6</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>PG0024-C</td>\n",
       "<td>ERS010150</td>\n",
       "<td>1041</td>\n",
       "<td>HB3xDD2 cross progeny</td>\n",
       "<td>3BD5</td>\n",
       "<td>progeny</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| sample      | acc                   | study  | study_title             | clone  | parent_or_progeny |\n",
       "+=============+=======================+========+=========================+========+===================+\n",
       "| 'PG0008-CW' | 'ERS010540'           | '1041' | 'HB3xDD2 cross progeny' | 'DD2'  | 'parent'          |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0004-CW' | 'ERS010539'           | '1041' | 'HB3xDD2 cross progeny' | 'HB3'  | 'parent'          |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0023-C'  | 'ERS010581,ERS010543' | '1041' | 'HB3xDD2 cross progeny' | '1BB5' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0022-Cx' | 'ERS088714'           | '1041' | 'HB3xDD2 cross progeny' | '3BA6' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "| 'PG0024-C'  | 'ERS010150'           | '1041' | 'HB3xDD2 cross progeny' | '3BD5' | 'progeny'         |\n",
       "+-------------+-----------------------+--------+-------------------------+--------+-------------------+\n",
       "..."
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_release5_crosses_metadata = etl.fromtsv(release5_crosses_metadata_txt_fn)\n",
    "print(len(tbl_release5_crosses_metadata.data()))\n",
    "tbl_release5_crosses_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "replicates_first = [\n",
    " 'PG0112-C',\n",
    " 'PG0062-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0055-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0004-CW',\n",
    " 'PG0111-C',\n",
    " 'PG0100-C',\n",
    " 'PG0079-C',\n",
    " 'PG0104-C',\n",
    " 'PG0086-C',\n",
    " 'PG0095-C',\n",
    " 'PG0078-C',\n",
    " 'PG0105-C',\n",
    " 'PG0102-C']\n",
    "\n",
    "replicates_second = [\n",
    " 'PG0112-CW',\n",
    " 'PG0065-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0067-C',\n",
    " 'PG0052-C',\n",
    " 'PG0111-CW',\n",
    " 'PG0100-CW',\n",
    " 'PG0079-CW',\n",
    " 'PG0104-CW',\n",
    " 'PG0086-CW',\n",
    " 'PG0095-CW',\n",
    " 'PG0078-CW',\n",
    " 'PG0105-CW',\n",
    " 'PG0102-CW']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "quads_first = [\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0053-C',\n",
    " 'PG0055-C',\n",
    " 'PG0055-C',\n",
    " 'PG0056-C']\n",
    "\n",
    "quads_second = [\n",
    " 'PG0055-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0056-C',\n",
    " 'PG0067-C',\n",
    " 'PG0067-C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "analysis_vcf_fn = \"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir)\n",
    "vcf_reader = vcf.Reader(filename=analysis_vcf_fn)\n",
    "sample_ids = np.array(vcf_reader.samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n",
      "[96 48 40 40 40 42 42 43  2 94 82 63 87 69 76 61 89 84]\n",
      "18\n",
      "[97 51 42 43 53 43 53 53 39 95 83 64 88 70 77 62 90 85]\n"
     ]
    }
   ],
   "source": [
    "# sample_ids = tbl_release5_crosses_metadata.values('sample').array()\n",
    "\n",
    "rep_index_first = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], replicates_first)]\n",
    "rep_index_second = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], replicates_second)]\n",
    "print(len(rep_index_first))\n",
    "print(rep_index_first)\n",
    "print(len(rep_index_second))\n",
    "print(rep_index_second)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n",
      "[40 40 40 42 42 43]\n",
      "6\n",
      "[42 43 53 43 53 53]\n"
     ]
    }
   ],
   "source": [
    "# sample_ids = tbl_release5_crosses_metadata.values('sample').array()\n",
    "\n",
    "quad_index_first = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], quads_first)]\n",
    "quad_index_second = np.argsort(sample_ids)[np.searchsorted(sample_ids[np.argsort(sample_ids)], quads_second)]\n",
    "print(len(quad_index_first))\n",
    "print(quad_index_first)\n",
    "print(len(quad_index_second))\n",
    "print(quad_index_second)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3D7xHB3 cross progeny', '7G8xGB4 cross progeny',\n",
       "       'HB3xDD2 cross progeny'], \n",
       "      dtype='<U21')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tbl_release5_crosses_metadata.distinct('study_title').values('study_title').array()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_variants_npy(vcf_fn):\n",
    "    output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    vcfnp.variants(\n",
    "        vcf_fn,\n",
    "        fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE', 'RU',\n",
    "                'SNPEFF_EFFECT', 'AC', 'AN', 'RPA', 'CDS', 'MULTIALLELIC',\n",
    "                'VQSLOD', 'FILTER'],\n",
    "        dtypes={\n",
    "            'REF':                      'a10',\n",
    "            'ALT':                      'a10',\n",
    "            'RegionType':               'a25',\n",
    "            'VariantType':              'a40',\n",
    "            'VARIANT_TYPE':             'a3',\n",
    "            'RU':                       'a40',\n",
    "            'SNPEFF_EFFECT':            'a33',\n",
    "            'CDS':                      bool,\n",
    "            'MULTIALLELIC':             bool,\n",
    "        },\n",
    "        arities={\n",
    "            'ALT':   1,\n",
    "            'AF':    1,\n",
    "            'AC':    1,\n",
    "            'RPA':   2,\n",
    "            'ANN':   1,\n",
    "        },\n",
    "        fills={\n",
    "            'VQSLOD': np.nan,\n",
    "        },\n",
    "        flatten_filter=True,\n",
    "        progress=100000,\n",
    "        verbose=True,\n",
    "        cache=True,\n",
    "        cachedir=output_dir\n",
    "    )\n",
    "\n",
    "def create_calldata_npy(vcf_fn, max_alleles=2):\n",
    "    output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    vcfnp.calldata_2d(\n",
    "        vcf_fn,\n",
    "        fields=['GT', 'AD', 'DP', 'GQ'],\n",
    "        dtypes={\n",
    "            'AD': 'u2',\n",
    "        },\n",
    "        arities={\n",
    "            'AD': max_alleles,\n",
    "        },\n",
    "        progress=100000,\n",
    "        verbose=True,\n",
    "        cache=True,\n",
    "        cachedir=output_dir\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[vcfnp] 2016-07-20 11:48:24.968050 :: caching is enabled\n",
      "[vcfnp] 2016-07-20 11:48:24.971688 :: cache file available\n",
      "[vcfnp] 2016-07-20 11:48:24.973228 :: loading from cache file /lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.analysis.vcf.gz.vcfnp_cache/variants.npy\n"
     ]
    }
   ],
   "source": [
    "# create_analysis_vcf(region='Pf3D7_14_v3', rewrite=True)\n",
    "# create_variants_npy(\"%s/vcf/SNP_INDEL_Pf3D7_14_v3:1000000-1100000.coding.vcf.gz\" % (output_dir))\n",
    "# create_calldata_npy(\"%s/vcf/SNP_INDEL_Pf3D7_14_v3:1000000-1100000.coding.vcf.gz\" % (output_dir))\n",
    "create_variants_npy(\"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir))\n",
    "create_calldata_npy(\"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "analysis_vcf_fn = \"%s/vcf/SNP_INDEL_WG.analysis.vcf.gz\" % (output_dir)\n",
    "variants = np.load(\"%s.vcfnp_cache/variants.npy\" % analysis_vcf_fn)\n",
    "calldata = np.load(\"%s.vcfnp_cache/calldata_2d.npy\" % analysis_vcf_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def genotype_concordance_gatk(calldata=calldata,\n",
    "                              ix = ((variants['VARIANT_TYPE'] == b'SNP') & (variants['MULTIALLELIC'] == False) &\n",
    "                                    (variants['CDS']) & variants['FILTER_PASS']),\n",
    "                              GQ_threshold=30,\n",
    "                              rep_index_first=rep_index_first, rep_index_second=rep_index_second,\n",
    "                              verbose=False):\n",
    "    GT = calldata['GT'][ix, :]\n",
    "    GT[calldata['GQ'][ix, :] < GQ_threshold] = b'./.'\n",
    "    \n",
    "    all_samples = sample_ids\n",
    "#     all_samples = tbl_release5_crosses_metadata.values('sample').array()\n",
    "    total_mendelian_errors = 0\n",
    "    total_homozygotes = 0\n",
    "    for cross in tbl_release5_crosses_metadata.distinct('study_title').values('study_title').array():\n",
    "        parents = (tbl_release5_crosses_metadata\n",
    "            .selecteq('study_title', cross)\n",
    "            .selecteq('parent_or_progeny', 'parent')\n",
    "            .values('sample')\n",
    "            .array()\n",
    "        )\n",
    "        parent1_calls = GT[:, all_samples==parents[0]]\n",
    "#         print(parent1_calls.shape)\n",
    "#         print(parent1_calls == b'1')\n",
    "        parent2_calls = GT[:, all_samples==parents[1]]\n",
    "#         print(parent1_calls.shape)\n",
    "        progeny = (tbl_release5_crosses_metadata\n",
    "            .selecteq('study_title', cross)\n",
    "            .selecteq('parent_or_progeny', 'progeny')\n",
    "            .values('sample')\n",
    "            .array()\n",
    "        )\n",
    "        mendelian_errors = np.zeros(len(progeny))\n",
    "        homozygotes = np.zeros(len(progeny))\n",
    "        for i, ox_code in enumerate(progeny):\n",
    "            progeny_calls = GT[:, all_samples==ox_code]\n",
    "#             print(ox_code, progeny_calls.shape)\n",
    "            error_calls = (\n",
    "                ((parent1_calls == b'0/0') & (parent2_calls == b'0/0') & (progeny_calls == b'1/1')) |\n",
    "                ((parent1_calls == b'1/1') & (parent2_calls == b'1/1') & (progeny_calls == b'0/0'))\n",
    "            )\n",
    "            homozygous_calls = (\n",
    "                ((parent1_calls == b'0/0') | (parent1_calls == b'1/1' )) &\n",
    "                ((parent2_calls == b'0/0') | (parent2_calls == b'1/1' )) &\n",
    "                ((progeny_calls == b'0/0') | (progeny_calls == b'1/1' ))\n",
    "            )\n",
    "            mendelian_errors[i] = np.sum(error_calls)\n",
    "            homozygotes[i] = np.sum(homozygous_calls)\n",
    "#         print(cross, mendelian_errors, homozygotes)\n",
    "        total_mendelian_errors = total_mendelian_errors + np.sum(mendelian_errors)\n",
    "        total_homozygotes = total_homozygotes + np.sum(homozygotes)\n",
    "        \n",
    "    mendelian_error_rate = total_mendelian_errors / total_homozygotes\n",
    "    \n",
    "    GT_both = (np.in1d(GT[:, rep_index_first], [b'0/0', b'1/1']) &\n",
    "                     np.in1d(GT[:, rep_index_second], [b'0/0', b'1/1'])\n",
    "                    )\n",
    "    GT_both = (\n",
    "        ((GT[:, rep_index_first] == b'0/0') | (GT[:, rep_index_first] == b'1/1')) &\n",
    "        ((GT[:, rep_index_second] == b'0/0') | (GT[:, rep_index_second] == b'1/1'))\n",
    "    )\n",
    "    GT_discordant = (\n",
    "        ((GT[:, rep_index_first] == b'0/0') & (GT[:, rep_index_second] == b'1/1')) |\n",
    "        ((GT[:, rep_index_first] == b'1/1') & (GT[:, rep_index_second] == b'0/0'))\n",
    "    )\n",
    "    missingness_per_sample = np.sum(GT == b'./.', 0)\n",
    "    if verbose:\n",
    "        print(missingness_per_sample)\n",
    "    mean_missingness = np.sum(missingness_per_sample) / (GT.shape[0] * GT.shape[1])\n",
    "    heterozygosity_per_sample = np.sum(GT == b'0/1', 0)\n",
    "    if verbose:\n",
    "        print(heterozygosity_per_sample)\n",
    "#     print(heterozygosity_per_sample)\n",
    "    mean_heterozygosity = np.sum(heterozygosity_per_sample) / (GT.shape[0] * GT.shape[1])\n",
    "#     print(calldata_both.shape)\n",
    "#     print(calldata_discordant.shape)\n",
    "    num_discordance_per_sample_pair = np.sum(GT_discordant, 0)\n",
    "    num_both_calls_per_sample_pair = np.sum(GT_both, 0)\n",
    "    if verbose:\n",
    "        print(num_discordance_per_sample_pair)\n",
    "        print(num_both_calls_per_sample_pair)\n",
    "    prop_discordances_per_sample_pair = num_discordance_per_sample_pair/num_both_calls_per_sample_pair\n",
    "    num_discordance_per_snp = np.sum(GT_discordant, 1)\n",
    "#     print(num_discordance_per_snp)\n",
    "#     print(variants_SNP_BIALLELIC[num_discordance_per_snp > 0])\n",
    "    mean_prop_discordances = (np.sum(num_discordance_per_sample_pair) / np.sum(num_both_calls_per_sample_pair))\n",
    "    return(\n",
    "        mean_missingness,\n",
    "        mean_heterozygosity,\n",
    "        mean_prop_discordances,\n",
    "        mendelian_error_rate,\n",
    "#         prop_discordances_per_sample_pair,\n",
    "#         GT.shape\n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.052356625919844309,\n",
       " 0.0032449761513800922,\n",
       " 1.4812620352540365e-05,\n",
       " 3.1193964068578418e-05)"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.017184907168486807,\n",
       " 0.0042406234524461125,\n",
       " 1.4499548097417631e-05,\n",
       " 3.1662086857739367e-05)"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.017184907168486807,\n",
       " 0.0042406234524461125,\n",
       " 1.4499548097417631e-05,\n",
       " 3.1662086857739367e-05)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_concordance_gatk()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30 True b'SNP' True False\n",
      "30 True b'SNP' True True\n",
      "30 True b'SNP' False False\n",
      "30 True b'SNP' False True\n",
      "30 True b'IND' True False\n",
      "30 True b'IND' True True\n",
      "30 True b'IND' False False\n",
      "30 True b'IND' False True\n",
      "30 True b'STR' True False\n",
      "30 True b'STR' True True\n",
      "30 True b'STR' False False\n",
      "30 True b'STR' False True\n",
      "30 False b'SNP' True False\n",
      "30 False b'SNP' True True\n",
      "30 False b'SNP' False False\n",
      "30 False b'SNP' False True\n",
      "30 False b'IND' True False\n",
      "30 False b'IND' True True\n",
      "30 False b'IND' False False\n",
      "30 False b'IND' False True\n",
      "30 False b'STR' True False\n",
      "30 False b'STR' True True\n",
      "30 False b'STR' False False\n",
      "30 False b'STR' False True\n",
      "99 True b'SNP' True False\n",
      "99 True b'SNP' True True\n",
      "99 True b'SNP' False False\n",
      "99 True b'SNP' False True\n",
      "99 True b'IND' True False\n",
      "99 True b'IND' True True\n",
      "99 True b'IND' False False\n",
      "99 True b'IND' False True\n",
      "99 True b'STR' True False\n",
      "99 True b'STR' True True\n",
      "99 True b'STR' False False\n",
      "99 True b'STR' False True\n",
      "99 False b'SNP' True False\n",
      "99 False b'SNP' True True\n",
      "99 False b'SNP' False False\n",
      "99 False b'SNP' False True\n",
      "99 False b'IND' True False\n",
      "99 False b'IND' True True\n",
      "99 False b'IND' False False\n",
      "99 False b'IND' False True\n",
      "99 False b'STR' True False\n",
      "99 False b'STR' True True\n",
      "99 False b'STR' False False\n",
      "99 False b'STR' False True\n",
      "0 True b'SNP' True False\n",
      "0 True b'SNP' True True\n",
      "0 True b'SNP' False False\n",
      "0 True b'SNP' False True\n",
      "0 True b'IND' True False\n",
      "0 True b'IND' True True\n",
      "0 True b'IND' False False\n",
      "0 True b'IND' False True\n",
      "0 True b'STR' True False\n",
      "0 True b'STR' True True\n",
      "0 True b'STR' False False\n",
      "0 True b'STR' False True\n",
      "0 False b'SNP' True False\n",
      "0 False b'SNP' True True\n",
      "0 False b'SNP' False False\n",
      "0 False b'SNP' False True\n",
      "0 False b'IND' True False\n",
      "0 False b'IND' True True\n",
      "0 False b'IND' False False\n",
      "0 False b'IND' False True\n",
      "0 False b'STR' True False\n",
      "0 False b'STR' True True\n",
      "0 False b'STR' False False\n",
      "0 False b'STR' False True\n",
      "[[30, True, b'SNP', True, False, 11745, 0.3847135036224486, 0.017184907168486807, 0.0042406234524461125, 1.4499548097417631e-05, 3.1662086857739367e-05], [30, True, b'SNP', True, True, 10006, 0.21235769535457111, 0.20619465055660482, 0.016412601418740593, 0.0023271800306471709, 0.0025315008109119642], [30, True, b'SNP', False, False, 9538, 0.34424388322127492, 0.095690278627701866, 0.0065580855953201161, 0.00010370620033445249, 0.00017185107554826832], [30, True, b'SNP', False, True, 26874, 0.099247956702417831, 0.30512231684368324, 0.016491168916774122, 0.0021986505991742475, 0.001170246727778538], [30, True, b'IND', True, False, 87, 0.23339831850859022, 0.082688247712878249, 0.020408163265306121, 0.0014513788098693759, 0.00030731407498463427], [30, True, b'IND', True, True, 5370, 0.14854666390536869, 0.17997377722038535, 0.011916163113290008, 0.0074118149245236652, 0.0068627868882407033], [30, True, b'IND', False, False, 613, 0.19350627040027488, 0.16849219296201351, 0.011818756866531279, 0.001006036217303823, 0.00088103301120563866], [30, True, b'IND', False, True, 30653, 0.085207204301829112, 0.31951761554783398, 0.015763347063942206, 0.0026752322906943402, 0.001688862039249829], [30, True, b'STR', True, False, 200, 0.25184482935328484, 0.053316326530612246, 0.0061224489795918364, 0.0006105006105006105, 0.00059520266650794592], [30, True, b'STR', True, True, 11981, 0.21511509790745284, 0.14807032904139036, 0.013440498476329018, 0.0038634141744896878, 0.0023771073899782245], [30, True, b'STR', False, False, 1316, 0.28281252468443419, 0.14649370386452454, 0.0070249984492277151, 0.00025225770647293276, 0.00032134440979743399], [30, True, b'STR', False, True, 155318, 0.13497247443958715, 0.39919798512124305, 0.022288899850234845, 0.001806925760681971, 0.00064932385027032229], [30, False, b'SNP', True, False, 14316, 0.13221499142215398, 0.41895609878486179, 0.080144379629471241, 0.00017495188823073656, 0.00030244535968759176], [30, False, b'SNP', True, True, 109329, 0.12746418712589855, 0.59057971623190886, 0.050438845790490829, 0.00053475379786603529, 0.0011007726006494229], [30, False, b'SNP', False, False, 43798, 0.086206690142186335, 0.4721376243999586, 0.061563476479682702, 0.00010411615793528734, 0.00027179365765253089], [30, False, b'SNP', False, True, 189502, 0.084234391715486912, 0.52492435059109821, 0.067898965688585694, 0.00045091758018007028, 0.00056162157539128203], [30, False, b'IND', True, False, 660, 0.069361678279875216, 0.32735003092145948, 0.043212739641311067, 0.001847640704945992, 0.0012824623276691247], [30, False, b'IND', True, True, 55969, 0.07913138975336885, 0.58389811998697527, 0.029109408597543612, 0.0031116701301542164, 0.0030757044853464365], [30, False, b'IND', False, False, 2670, 0.04156587965153985, 0.32757012917526562, 0.030340900405105863, 0.0010196362201098576, 0.00089930666357227818], [30, False, b'IND', False, True, 73442, 0.043558096698742194, 0.43354105891696293, 0.035154354762247479, 0.00082155043249436562, 0.00067049142982701323], [30, False, b'STR', True, False, 445, 0.03217216064488268, 0.15010318734235267, 0.022127952304517313, 0.0003348961821835231, 3.2654127481713692e-05], [30, False, b'STR', True, True, 13768, 0.058442608937037767, 0.47972376050943327, 0.0352451410546787, 0.0020318284328326454, 0.0012359647288871399], [30, False, b'STR', False, False, 986, 0.083607318542315026, 0.43661257606490872, 0.039036304176843152, 0.00076603195447581525, 0.00033089451818081546], [30, False, b'STR', False, True, 64157, 0.058904449615435554, 0.54917019569022807, 0.041887837012074651, 0.0021364552749142487, 0.0010306599209311327], [99, True, b'SNP', True, False, 11745, 0.3847135036224486, 0.052356625919844309, 0.0032449761513800922, 1.4812620352540365e-05, 3.1193964068578418e-05], [99, True, b'SNP', True, True, 10006, 0.21235769535457111, 0.29673216478276299, 0.013492924653371242, 0.0009533570083657077, 0.0013091052800579628], [99, True, b'SNP', False, False, 9538, 0.34424388322127492, 0.26537352202361336, 0.004240824029339142, 0.00011577232545254575, 0.0001457538112509268], [99, True, b'SNP', False, True, 26874, 0.099247956702417831, 0.61375420898433053, 0.0084616342629929842, 0.00083804896051174803, 0.00058613436171611375], [99, True, b'IND', True, False, 87, 0.23339831850859022, 0.12420830401125968, 0.017593244194229415, 0.0015117157974300832, 0.00032578595862518323], [99, True, b'IND', True, True, 5370, 0.14854666390536869, 0.30590582601755784, 0.0087542279481625049, 0.0040584415584415581, 0.0043556909601561384], [99, True, b'IND', False, False, 613, 0.19350627040027488, 0.41517128874388254, 0.006641808436261944, 0.00034311202607651396, 0.0010031767262999498], [99, True, b'IND', False, True, 30653, 0.085207204301829112, 0.64923431937613718, 0.0066391610635706992, 0.00095344402585926234, 0.00084157860344156142], [99, True, b'STR', True, False, 200, 0.25184482935328484, 0.13750000000000001, 0.0033163265306122448, 0.0, 0.00040477636106051409], [99, True, b'STR', True, True, 11981, 0.21511509790745284, 0.29378062885282652, 0.0060955356184707416, 0.0024912378382622724, 0.0015081208313300748], [99, True, b'STR', False, False, 1316, 0.28281252468443419, 0.37448049128465977, 0.0031635754605793685, 0.00015542430836182778, 0.00024637161798960757], [99, True, b'STR', False, True, 155318, 0.13497247443958715, 0.72634109980025185, 0.0058329310425930632, 0.00043725994563941167, 0.00023198556534260089], [99, False, b'SNP', True, False, 14316, 0.13221499142215398, 0.58095480438613001, 0.064069173352492709, 1.1316313597682419e-05, 9.995102399824087e-06], [99, False, b'SNP', True, True, 109329, 0.12746418712589855, 0.72966692370771535, 0.038438370161883596, 4.8037360535010884e-05, 0.00027815373127362673], [99, False, b'SNP', False, False, 43798, 0.086206690142186335, 0.67217797662925616, 0.045260430305735698, 3.9519834016697127e-05, 6.2386612332086439e-05], [99, False, b'SNP', False, True, 189502, 0.084234391715486912, 0.71324571664635927, 0.049116707399997288, 4.4075633787579487e-05, 0.00010920639130405108], [99, False, b'IND', True, False, 660, 0.069361678279875216, 0.48070500927643783, 0.03688930117501546, 0.00053763440860215054, 0.00035627338044059141], [99, False, b'IND', True, True, 55969, 0.07913138975336885, 0.72916658310486016, 0.022466883088706903, 0.00060593348714721852, 0.0010333455024719852], [99, False, b'IND', False, False, 2670, 0.04156587965153985, 0.56503859970954673, 0.02208209126347168, 0.00077211559673505403, 0.00056734906151009409], [99, False, b'IND', False, True, 73442, 0.043558096698742194, 0.65853451481079894, 0.024671697060404185, 0.00015746919848177974, 0.00012031188542606604], [99, False, b'STR', True, False, 445, 0.03217216064488268, 0.29605595047007566, 0.014102270121531759, 0.00023719165085388995, 0.0], [99, False, b'STR', True, True, 13768, 0.058442608937037767, 0.68419375303869367, 0.020150986019044456, 0.00053988392495613448, 0.00039040999449585907], [99, False, b'STR', False, False, 986, 0.083607318542315026, 0.64563066605952724, 0.030384567620151508, 0.00018768768768768769, 9.0415913200723327e-05], [99, False, b'STR', False, True, 64157, 0.058904449615435554, 0.79717644184721603, 0.021478401357893407, 0.00020145044319097501, 0.00022661096882602692], [0, True, b'SNP', True, False, 11745, 0.3847135036224486, 0.0086280744737230776, 0.0049287147809315294, 4.7977277961157595e-05, 8.8130357117473978e-05], [0, True, b'SNP', True, True, 10006, 0.21235769535457111, 0.083431573708835929, 0.020763052372658037, 0.007289836888331242, 0.0061039732670332657], [0, True, b'SNP', False, False, 9538, 0.34424388322127492, 0.01696543578639256, 0.0084452736850663939, 0.00029371744381904606, 0.00026735263089440296], [0, True, b'SNP', False, True, 26874, 0.099247956702417831, 0.045450575854364968, 0.025906991508369368, 0.0052491628688375773, 0.0035236284870591322], [0, True, b'IND', True, False, 87, 0.23339831850859022, 0.037414965986394558, 0.023105794041754633, 0.0013783597518952446, 0.0016968325791855204], [0, True, b'IND', True, True, 5370, 0.14854666390536869, 0.064108995553528669, 0.017787785505263557, 0.013219178082191781, 0.011404446241683757], [0, True, b'IND', False, False, 613, 0.19350627040027488, 0.03102839830875254, 0.016979059160368879, 0.0042368801155512762, 0.0023845481281297193], [0, True, b'IND', False, True, 30653, 0.085207204301829112, 0.050350966080491506, 0.028527686806298549, 0.0083011456360042972, 0.0049298280673706653], [0, True, b'STR', True, False, 200, 0.25184482935328484, 0.004387755102040816, 0.0095408163265306122, 0.0011366865586814436, 0.00066692602678819545], [0, True, b'STR', True, True, 11981, 0.21511509790745284, 0.019023317531670043, 0.022510982525052421, 0.007921492088465789, 0.0047554206752519366], [0, True, b'STR', False, False, 1316, 0.28281252468443419, 0.018376651572483096, 0.012553501643818622, 0.00175054704595186, 0.001402101377254011], [0, True, b'STR', False, True, 155318, 0.13497247443958715, 0.026305084157821308, 0.044560652523026489, 0.010794456242502979, 0.0060259110587351618], [0, False, b'SNP', True, False, 14316, 0.13221499142215398, 0.17777668485667528, 0.087320594625109049, 0.0058670585575401309, 0.0055263868648884485], [0, False, b'SNP', True, True, 109329, 0.12746418712589855, 0.3397775596257766, 0.055871241287997787, 0.010152916769366077, 0.010053769480430763], [0, False, b'SNP', False, False, 43798, 0.086206690142186335, 0.1437224325777619, 0.070539983654085411, 0.0030107439582149334, 0.002656988541966809], [0, False, b'SNP', False, True, 189502, 0.084234391715486912, 0.1781470617185883, 0.077090565411080683, 0.0077880249899755531, 0.0051229363885763289], [0, False, b'IND', True, False, 660, 0.069361678279875216, 0.17237167594310451, 0.045423623995052564, 0.0068344913999316553, 0.0045836306746445624], [0, False, b'IND', True, True, 55969, 0.07913138975336885, 0.35072859210328167, 0.03217943898243962, 0.013286776551805728, 0.010349964199480045], [0, False, b'IND', False, False, 2670, 0.04156587965153985, 0.12786058243522128, 0.035576702591148819, 0.0049083518674744994, 0.002953767374670417], [0, False, b'IND', False, True, 73442, 0.043558096698742194, 0.15132432701301429, 0.041063918827518482, 0.0069998566342740331, 0.0041655558115716316], [0, False, b'STR', True, False, 445, 0.03217216064488268, 0.044210043567988994, 0.031414813116257739, 0.0028145229383619475, 0.0015017520440513933], [0, False, b'STR', True, True, 13768, 0.058442608937037767, 0.19369226481993146, 0.048781409716704811, 0.0076701486891494744, 0.0040873292792343733], [0, False, b'STR', False, False, 986, 0.083607318542315026, 0.17531150391190958, 0.045131845841784993, 0.0050689526649273205, 0.0031159942363112392], [0, False, b'STR', False, True, 64157, 0.058904449615435554, 0.093326702066645825, 0.063844815635623456, 0.0094323709946772562, 0.0044767741854289632]]\n"
     ]
    }
   ],
   "source": [
    "results_list = list()\n",
    "GQ_thresholds = [30, 99, 0]\n",
    "variant_types = [b'SNP', b'IND', b'STR']\n",
    "multiallelics = [False, True]\n",
    "codings = [True, False]\n",
    "filter_passes = [True, False]\n",
    "\n",
    "for GQ_threshold in GQ_thresholds:\n",
    "    for filter_pass in filter_passes:\n",
    "        for variant_type in variant_types:\n",
    "            for coding in codings:\n",
    "                for multiallelic in multiallelics:\n",
    "                    print(GQ_threshold, filter_pass, variant_type, coding, multiallelic)\n",
    "                    ix = (\n",
    "                        (variants['VARIANT_TYPE'] == variant_type) &\n",
    "                        (variants['MULTIALLELIC'] == multiallelic) &\n",
    "                        (variants['CDS'] == coding) &\n",
    "                        (variants['FILTER_PASS'] == filter_pass)\n",
    "                    )\n",
    "                    number_of_alleles = np.sum(ix)\n",
    "                    mean_nraf = np.sum(variants['AC'][ix]) / np.sum(variants['AN'][ix])\n",
    "                    genotype_quality_results = list(genotype_concordance_gatk(ix=ix, GQ_threshold=GQ_threshold))\n",
    "                    results_list.append(\n",
    "                        [GQ_threshold, filter_pass, variant_type, coding, multiallelic, number_of_alleles, mean_nraf] +\n",
    "                        genotype_quality_results\n",
    "                    )\n",
    "\n",
    "print(results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30 True SNP True False\n",
      "30 True SNP True True\n",
      "30 True SNP False False\n",
      "30 True SNP False True\n",
      "30 True indel True False\n",
      "30 True indel True True\n",
      "30 True indel False False\n",
      "30 True indel False True\n",
      "30 False SNP True False\n",
      "30 False SNP True True\n",
      "30 False SNP False False\n",
      "30 False SNP False True\n",
      "30 False indel True False\n",
      "30 False indel True True\n",
      "30 False indel False False\n",
      "30 False indel False True\n",
      "99 True SNP True False\n",
      "99 True SNP True True\n",
      "99 True SNP False False\n",
      "99 True SNP False True\n",
      "99 True indel True False\n",
      "99 True indel True True\n",
      "99 True indel False False\n",
      "99 True indel False True\n",
      "99 False SNP True False\n",
      "99 False SNP True True\n",
      "99 False SNP False False\n",
      "99 False SNP False True\n",
      "99 False indel True False\n",
      "99 False indel True True\n",
      "99 False indel False False\n",
      "99 False indel False True\n",
      "0 True SNP True False\n",
      "0 True SNP True True\n",
      "0 True SNP False False\n",
      "0 True SNP False True\n",
      "0 True indel True False\n",
      "0 True indel True True\n",
      "0 True indel False False\n",
      "0 True indel False True\n",
      "0 False SNP True False\n",
      "0 False SNP True True\n",
      "0 False SNP False False\n",
      "0 False SNP False True\n",
      "0 False indel True False\n",
      "0 False indel True True\n",
      "0 False indel False False\n",
      "0 False indel False True\n"
     ]
    }
   ],
   "source": [
    "results_list = list()\n",
    "GQ_thresholds = [30, 99, 0]\n",
    "variant_types = collections.OrderedDict()\n",
    "variant_types['SNP'] = [b'SNP']\n",
    "variant_types['indel'] = [b'IND', b'STR']\n",
    "multiallelics = [False, True]\n",
    "codings = [True, False]\n",
    "filter_passes = [True, False]\n",
    "\n",
    "for GQ_threshold in GQ_thresholds:\n",
    "    for filter_pass in filter_passes:\n",
    "        for variant_type in variant_types:\n",
    "            for coding in codings:\n",
    "                for multiallelic in multiallelics:\n",
    "                    print(GQ_threshold, filter_pass, variant_type, coding, multiallelic)\n",
    "                    ix = (\n",
    "                        (np.in1d(variants['VARIANT_TYPE'], variant_types[variant_type])) &\n",
    "                        (variants['MULTIALLELIC'] == multiallelic) &\n",
    "                        (variants['CDS'] == coding) &\n",
    "                        (variants['FILTER_PASS'] == filter_pass)\n",
    "                    )\n",
    "                    number_of_alleles = np.sum(ix)\n",
    "                    mean_nraf = np.sum(variants['AC'][ix]) / np.sum(variants['AN'][ix])\n",
    "                    genotype_quality_results = list(genotype_concordance_gatk(ix=ix, GQ_threshold=GQ_threshold))\n",
    "                    results_list.append(\n",
    "                        [GQ_threshold, filter_pass, variant_type, coding, multiallelic, number_of_alleles, mean_nraf] +\n",
    "                        genotype_quality_results\n",
    "                    )\n",
    "\n",
    "# print(results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>GQ threshold</th>\n",
       "<th>PASS</th>\n",
       "<th>Type</th>\n",
       "<th>Coding</th>\n",
       "<th>Multiallelic</th>\n",
       "<th>Alleles</th>\n",
       "<th>Mean NRAF</th>\n",
       "<th>Missingness</th>\n",
       "<th>Heterozygosity</th>\n",
       "<th>Discordance</th>\n",
       "<th>MER</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.0171849071685</td>\n",
       "<td style='text-align: right'>0.00424062345245</td>\n",
       "<td style='text-align: right'>1.44995480974e-05</td>\n",
       "<td style='text-align: right'>3.16620868577e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>10006</td>\n",
       "<td style='text-align: right'>0.212357695355</td>\n",
       "<td style='text-align: right'>0.206194650557</td>\n",
       "<td style='text-align: right'>0.0164126014187</td>\n",
       "<td style='text-align: right'>0.00232718003065</td>\n",
       "<td style='text-align: right'>0.00253150081091</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.0956902786277</td>\n",
       "<td style='text-align: right'>0.00655808559532</td>\n",
       "<td style='text-align: right'>0.000103706200334</td>\n",
       "<td style='text-align: right'>0.000171851075548</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>26874</td>\n",
       "<td style='text-align: right'>0.0992479567024</td>\n",
       "<td style='text-align: right'>0.305122316844</td>\n",
       "<td style='text-align: right'>0.0164911689168</td>\n",
       "<td style='text-align: right'>0.00219865059917</td>\n",
       "<td style='text-align: right'>0.00117024672778</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.0622200099552</td>\n",
       "<td style='text-align: right'>0.0104529616725</td>\n",
       "<td style='text-align: right'>0.000859475719811</td>\n",
       "<td style='text-align: right'>0.000514822600712</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>17351</td>\n",
       "<td style='text-align: right'>0.195175950725</td>\n",
       "<td style='text-align: right'>0.157944198946</td>\n",
       "<td style='text-align: right'>0.0129687284977</td>\n",
       "<td style='text-align: right'>0.0049735539665</td>\n",
       "<td style='text-align: right'>0.00367479637559</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.153484410872</td>\n",
       "<td style='text-align: right'>0.00854836491362</td>\n",
       "<td style='text-align: right'>0.00048666875239</td>\n",
       "<td style='text-align: right'>0.000490265324946</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>185971</td>\n",
       "<td style='text-align: right'>0.126939696755</td>\n",
       "<td style='text-align: right'>0.386064526848</td>\n",
       "<td style='text-align: right'>0.0212133140355</td>\n",
       "<td style='text-align: right'>0.0019951671589</td>\n",
       "<td style='text-align: right'>0.000862573285925</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.418956098785</td>\n",
       "<td style='text-align: right'>0.0801443796295</td>\n",
       "<td style='text-align: right'>0.000174951888231</td>\n",
       "<td style='text-align: right'>0.000302445359688</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>109329</td>\n",
       "<td style='text-align: right'>0.127464187126</td>\n",
       "<td style='text-align: right'>0.590579716232</td>\n",
       "<td style='text-align: right'>0.0504388457905</td>\n",
       "<td style='text-align: right'>0.000534753797866</td>\n",
       "<td style='text-align: right'>0.00110077260065</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.4721376244</td>\n",
       "<td style='text-align: right'>0.0615634764797</td>\n",
       "<td style='text-align: right'>0.000104116157935</td>\n",
       "<td style='text-align: right'>0.000271793657653</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>189502</td>\n",
       "<td style='text-align: right'>0.0842343917155</td>\n",
       "<td style='text-align: right'>0.524924350591</td>\n",
       "<td style='text-align: right'>0.0678989656886</td>\n",
       "<td style='text-align: right'>0.00045091758018</td>\n",
       "<td style='text-align: right'>0.000561621575391</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.25597008034</td>\n",
       "<td style='text-align: right'>0.0347215809401</td>\n",
       "<td style='text-align: right'>0.00115313653137</td>\n",
       "<td style='text-align: right'>0.000593781488412</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>69737</td>\n",
       "<td style='text-align: right'>0.0742901230784</td>\n",
       "<td style='text-align: right'>0.563331238973</td>\n",
       "<td style='text-align: right'>0.0303207707793</td>\n",
       "<td style='text-align: right'>0.00289086495375</td>\n",
       "<td style='text-align: right'>0.00251653549782</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.356978185147</td>\n",
       "<td style='text-align: right'>0.0326859956236</td>\n",
       "<td style='text-align: right'>0.000961027793935</td>\n",
       "<td style='text-align: right'>0.000780938964261</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>30</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>137599</td>\n",
       "<td style='text-align: right'>0.0509664324683</td>\n",
       "<td style='text-align: right'>0.487454376077</td>\n",
       "<td style='text-align: right'>0.038293912613</td>\n",
       "<td style='text-align: right'>0.00126552588647</td>\n",
       "<td style='text-align: right'>0.000788447594855</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.0523566259198</td>\n",
       "<td style='text-align: right'>0.00324497615138</td>\n",
       "<td style='text-align: right'>1.48126203525e-05</td>\n",
       "<td style='text-align: right'>3.11939640686e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>10006</td>\n",
       "<td style='text-align: right'>0.212357695355</td>\n",
       "<td style='text-align: right'>0.296732164783</td>\n",
       "<td style='text-align: right'>0.0134929246534</td>\n",
       "<td style='text-align: right'>0.000953357008366</td>\n",
       "<td style='text-align: right'>0.00130910528006</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.265373522024</td>\n",
       "<td style='text-align: right'>0.00424082402934</td>\n",
       "<td style='text-align: right'>0.000115772325453</td>\n",
       "<td style='text-align: right'>0.000145753811251</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>26874</td>\n",
       "<td style='text-align: right'>0.0992479567024</td>\n",
       "<td style='text-align: right'>0.613754208984</td>\n",
       "<td style='text-align: right'>0.00846163426299</td>\n",
       "<td style='text-align: right'>0.000838048960512</td>\n",
       "<td style='text-align: right'>0.000586134361716</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.133470809927</td>\n",
       "<td style='text-align: right'>0.00764417265164</td>\n",
       "<td style='text-align: right'>0.000468055230517</td>\n",
       "<td style='text-align: right'>0.000381642972999</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>17351</td>\n",
       "<td style='text-align: right'>0.195175950725</td>\n",
       "<td style='text-align: right'>0.297533283384</td>\n",
       "<td style='text-align: right'>0.00691838028509</td>\n",
       "<td style='text-align: right'>0.00299585226915</td>\n",
       "<td style='text-align: right'>0.00231410006658</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.387411263106</td>\n",
       "<td style='text-align: right'>0.00426889262704</td>\n",
       "<td style='text-align: right'>0.00021393806493</td>\n",
       "<td style='text-align: right'>0.00046332539822</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>185971</td>\n",
       "<td style='text-align: right'>0.126939696755</td>\n",
       "<td style='text-align: right'>0.713631837924</td>\n",
       "<td style='text-align: right'>0.00596581933611</td>\n",
       "<td style='text-align: right'>0.000570576282045</td>\n",
       "<td style='text-align: right'>0.000361326319681</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.580954804386</td>\n",
       "<td style='text-align: right'>0.0640691733525</td>\n",
       "<td style='text-align: right'>1.13163135977e-05</td>\n",
       "<td style='text-align: right'>9.99510239982e-06</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>109329</td>\n",
       "<td style='text-align: right'>0.127464187126</td>\n",
       "<td style='text-align: right'>0.729666923708</td>\n",
       "<td style='text-align: right'>0.0384383701619</td>\n",
       "<td style='text-align: right'>4.8037360535e-05</td>\n",
       "<td style='text-align: right'>0.000278153731274</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.672177976629</td>\n",
       "<td style='text-align: right'>0.0452604303057</td>\n",
       "<td style='text-align: right'>3.95198340167e-05</td>\n",
       "<td style='text-align: right'>6.23866123321e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>189502</td>\n",
       "<td style='text-align: right'>0.0842343917155</td>\n",
       "<td style='text-align: right'>0.713245716646</td>\n",
       "<td style='text-align: right'>0.0491167074</td>\n",
       "<td style='text-align: right'>4.40756337876e-05</td>\n",
       "<td style='text-align: right'>0.000109206391304</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.406344076092</td>\n",
       "<td style='text-align: right'>0.0277126235109</td>\n",
       "<td style='text-align: right'>0.000408329930584</td>\n",
       "<td style='text-align: right'>0.000146448620942</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>69737</td>\n",
       "<td style='text-align: right'>0.0742901230784</td>\n",
       "<td style='text-align: right'>0.720287710708</td>\n",
       "<td style='text-align: right'>0.022009661372</td>\n",
       "<td style='text-align: right'>0.000594858995787</td>\n",
       "<td style='text-align: right'>0.000823829744634</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.586773768588</td>\n",
       "<td style='text-align: right'>0.0243212164516</td>\n",
       "<td style='text-align: right'>0.000639386189258</td>\n",
       "<td style='text-align: right'>0.000468498182227</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>99</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>137599</td>\n",
       "<td style='text-align: right'>0.0509664324683</td>\n",
       "<td style='text-align: right'>0.723177790655</td>\n",
       "<td style='text-align: right'>0.0231827889115</td>\n",
       "<td style='text-align: right'>0.000169203390438</td>\n",
       "<td style='text-align: right'>0.000148493078636</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>11745</td>\n",
       "<td style='text-align: right'>0.384713503622</td>\n",
       "<td style='text-align: right'>0.00862807447372</td>\n",
       "<td style='text-align: right'>0.00492871478093</td>\n",
       "<td style='text-align: right'>4.79772779612e-05</td>\n",
       "<td style='text-align: right'>8.81303571175e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>10006</td>\n",
       "<td style='text-align: right'>0.212357695355</td>\n",
       "<td style='text-align: right'>0.0834315737088</td>\n",
       "<td style='text-align: right'>0.0207630523727</td>\n",
       "<td style='text-align: right'>0.00728983688833</td>\n",
       "<td style='text-align: right'>0.00610397326703</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>9538</td>\n",
       "<td style='text-align: right'>0.344243883221</td>\n",
       "<td style='text-align: right'>0.0169654357864</td>\n",
       "<td style='text-align: right'>0.00844527368507</td>\n",
       "<td style='text-align: right'>0.000293717443819</td>\n",
       "<td style='text-align: right'>0.000267352630894</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>26874</td>\n",
       "<td style='text-align: right'>0.0992479567024</td>\n",
       "<td style='text-align: right'>0.0454505758544</td>\n",
       "<td style='text-align: right'>0.0259069915084</td>\n",
       "<td style='text-align: right'>0.00524916286884</td>\n",
       "<td style='text-align: right'>0.00352362848706</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>287</td>\n",
       "<td style='text-align: right'>0.246383608095</td>\n",
       "<td style='text-align: right'>0.0143994880182</td>\n",
       "<td style='text-align: right'>0.0136528478987</td>\n",
       "<td style='text-align: right'>0.00120724346076</td>\n",
       "<td style='text-align: right'>0.000957510472771</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>17351</td>\n",
       "<td style='text-align: right'>0.195175950725</td>\n",
       "<td style='text-align: right'>0.0329769853881</td>\n",
       "<td style='text-align: right'>0.0210491896603</td>\n",
       "<td style='text-align: right'>0.00953038003647</td>\n",
       "<td style='text-align: right'>0.00669157696785</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>1929</td>\n",
       "<td style='text-align: right'>0.254683238821</td>\n",
       "<td style='text-align: right'>0.0223971392601</td>\n",
       "<td style='text-align: right'>0.0139598607717</td>\n",
       "<td style='text-align: right'>0.00252745599519</td>\n",
       "<td style='text-align: right'>0.00170539592177</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>True</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>185971</td>\n",
       "<td style='text-align: right'>0.126939696755</td>\n",
       "<td style='text-align: right'>0.0302684893047</td>\n",
       "<td style='text-align: right'>0.0419179905052</td>\n",
       "<td style='text-align: right'>0.0103717493568</td>\n",
       "<td style='text-align: right'>0.0058470965276</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>14316</td>\n",
       "<td style='text-align: right'>0.132214991422</td>\n",
       "<td style='text-align: right'>0.177776684857</td>\n",
       "<td style='text-align: right'>0.0873205946251</td>\n",
       "<td style='text-align: right'>0.00586705855754</td>\n",
       "<td style='text-align: right'>0.00552638686489</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>109329</td>\n",
       "<td style='text-align: right'>0.127464187126</td>\n",
       "<td style='text-align: right'>0.339777559626</td>\n",
       "<td style='text-align: right'>0.055871241288</td>\n",
       "<td style='text-align: right'>0.0101529167694</td>\n",
       "<td style='text-align: right'>0.0100537694804</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>43798</td>\n",
       "<td style='text-align: right'>0.0862066901422</td>\n",
       "<td style='text-align: right'>0.143722432578</td>\n",
       "<td style='text-align: right'>0.0705399836541</td>\n",
       "<td style='text-align: right'>0.00301074395821</td>\n",
       "<td style='text-align: right'>0.00265698854197</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>SNP</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>189502</td>\n",
       "<td style='text-align: right'>0.0842343917155</td>\n",
       "<td style='text-align: right'>0.178147061719</td>\n",
       "<td style='text-align: right'>0.0770905654111</td>\n",
       "<td style='text-align: right'>0.00778802498998</td>\n",
       "<td style='text-align: right'>0.00512293638858</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>1105</td>\n",
       "<td style='text-align: right'>0.05308098684</td>\n",
       "<td style='text-align: right'>0.12075907286</td>\n",
       "<td style='text-align: right'>0.0397820666728</td>\n",
       "<td style='text-align: right'>0.00503619767076</td>\n",
       "<td style='text-align: right'>0.00305282351641</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>69737</td>\n",
       "<td style='text-align: right'>0.0742901230784</td>\n",
       "<td style='text-align: right'>0.319725306128</td>\n",
       "<td style='text-align: right'>0.0354571241864</td>\n",
       "<td style='text-align: right'>0.0119941013934</td>\n",
       "<td style='text-align: right'>0.00851794132839</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>3656</td>\n",
       "<td style='text-align: right'>0.0524469619217</td>\n",
       "<td style='text-align: right'>0.140657794847</td>\n",
       "<td style='text-align: right'>0.0381536640915</td>\n",
       "<td style='text-align: right'>0.00494936419706</td>\n",
       "<td style='text-align: right'>0.00299365398497</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>False</td>\n",
       "<td>indel</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>137599</td>\n",
       "<td style='text-align: right'>0.0509664324683</td>\n",
       "<td style='text-align: right'>0.124282316361</td>\n",
       "<td style='text-align: right'>0.0516857547167</td>\n",
       "<td style='text-align: right'>0.00815041655571</td>\n",
       "<td style='text-align: right'>0.00431787258759</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "headers = ['GQ threshold', 'PASS', 'Type', 'Coding', 'Multiallelic', 'Alleles', 'Mean NRAF', 'Missingness',\n",
    "           'Heterozygosity', 'Discordance', 'MER']\n",
    "etl.wrap(results_list).pushheader(headers).displayall()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/genotype_quality.xlsx'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# etl.wrap(results_list).pushheader(headers).convert('Alleles', int).toxlsx(results_table_fn)\n",
    "etl.wrap(results_list).pushheader(headers).cutout('Alleles').cutout('Mean NRAF').toxlsx(results_table_fn)\n",
    "results_table_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "variant_type_header_2_fn = \"%s/vcf/VARIANT_TYPE_2.hdr\" % (output_dir)\n",
    "fo=open(variant_type_header_2_fn, 'w')\n",
    "print('##INFO=<ID=VARIANT_TYPE,Number=1,Type=String,Description=\"SNP or INDEL\">', file=fo)\n",
    "fo.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_variant_counts_vcf(input_vcf_fn=sites_only_vcf_fn, region='Pf3D7_14_v3:1000000-1100000',\n",
    "                        output_vcf_fn=None, BCFTOOLS=BCFTOOLS, rewrite=False):\n",
    "    if output_vcf_fn is None:\n",
    "        if region is None:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir)\n",
    "        else:\n",
    "            output_vcf_fn = \"%s/vcf/SNP_INDEL_%s.sites.analysis.vcf.gz\" % (output_dir, region)\n",
    "    intermediate_fns = collections.OrderedDict()\n",
    "    for intermediate_file in ['multiallelic', 'multiallelics', 'snps', 'indels',\n",
    "                              'variant_type', 'analysis']:\n",
    "        intermediate_fns[intermediate_file] = output_vcf_fn.replace('analysis', intermediate_file)\n",
    "    \n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelic']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\t1\\n' --include 'N_ALT>1' {input_vcf_fn} | bgzip -c > {intermediate_fns['multiallelic']} && tabix -s1 -b2 -e2 {intermediate_fns['multiallelic']}\n",
    "        \n",
    "    if rewrite or not os.path.exists(intermediate_fns['multiallelics']):\n",
    "        !{BCFTOOLS} annotate --remove INFO/AF,INFO/BaseQRankSum,INFO/ClippingRankSum,INFO/DP,INFO/DS,INFO/END,INFO/FS,\\\n",
    "INFO/GC,INFO/HaplotypeScore,INFO/InbreedingCoeff,INFO/MLEAC,INFO/MLEAF,INFO/MQ,INFO/MQRankSum,INFO/NEGATIVE_TRAIN_SITE,\\\n",
    "INFO/POSITIVE_TRAIN_SITE,INFO/QD,INFO/ReadPosRankSum,INFO/RegionType,INFO/RPA,INFO/RU,INFO/STR,INFO/SNPEFF_AMINO_ACID_CHANGE,\\\n",
    "INFO/SNPEFF_CODON_CHANGE,INFO/SNPEFF_EXON_ID,INFO/SNPEFF_FUNCTIONAL_CLASS,INFO/SNPEFF_GENE_BIOTYPE,\\\n",
    "INFO/SNPEFF_GENE_NAME,INFO/SNPEFF_IMPACT,INFO/SNPEFF_TRANSCRIPT_ID,INFO/SOR,INFO/culprit,INFO/set\\\n",
    "        -a {intermediate_fns['multiallelic']} -c CHROM,POS,INFO/MULTIALLELIC\\\n",
    "        -h {multiallelic_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['multiallelics']} {input_vcf_fn}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['multiallelics']}\n",
    "                \n",
    "    if rewrite or not os.path.exists(intermediate_fns['snps']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tSNP\\n' --include 'TYPE=\"snp\"' {intermediate_fns['multiallelics']} | bgzip -c > {intermediate_fns['snps']} && tabix -s1 -b2 -e2 -f {intermediate_fns['snps']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['indels']):\n",
    "        !{BCFTOOLS} query -f'%CHROM\\t%POS\\tINDEL\\n' --include 'TYPE!=\"snp\"' {intermediate_fns['multiallelics']} | bgzip -c > {intermediate_fns['indels']} && tabix -s1 -b2 -e2 -f {intermediate_fns['indels']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['variant_type']):\n",
    "        !{BCFTOOLS} annotate -a {intermediate_fns['snps']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -h {variant_type_header_2_fn} {intermediate_fns['multiallelics']} | \\\n",
    "        {BCFTOOLS} annotate -a {intermediate_fns['indels']} -c CHROM,POS,INFO/VARIANT_TYPE \\\n",
    "        -Oz -o {intermediate_fns['variant_type']} \n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['variant_type']}\n",
    "\n",
    "    if rewrite or not os.path.exists(intermediate_fns['analysis']):\n",
    "        !{BCFTOOLS} annotate -a {cds_gff_fn} -c CHROM,FROM,TO,CDS \\\n",
    "        -h {cds_header_fn} \\\n",
    "        -Oz -o {intermediate_fns['analysis']} {intermediate_fns['variant_type']}\n",
    "        !{BCFTOOLS} index --tbi {intermediate_fns['analysis']}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "create_variant_counts_vcf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "create_variant_counts_vcf(region=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def create_variants_npy_2(vcf_fn):\n",
    "    output_dir = '%s.vcfnp_cache' % vcf_fn\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    vcfnp.variants(\n",
    "        vcf_fn,\n",
    "        fields=['CHROM', 'POS', 'REF', 'ALT', 'VariantType', 'VARIANT_TYPE',\n",
    "                'SNPEFF_EFFECT', 'AC', 'AN', 'CDS', 'MULTIALLELIC',\n",
    "                'VQSLOD', 'FILTER'],\n",
    "        dtypes={\n",
    "            'REF':                      'a10',\n",
    "            'ALT':                      'a10',\n",
    "            'RegionType':               'a25',\n",
    "            'VariantType':              'a40',\n",
    "            'VARIANT_TYPE':             'a3',\n",
    "            'SNPEFF_EFFECT':            'a33',\n",
    "            'CDS':                      bool,\n",
    "            'MULTIALLELIC':             bool,\n",
    "        },\n",
    "        arities={\n",
    "            'ALT':   1,\n",
    "            'AF':    1,\n",
    "            'AC':    1,\n",
    "            'ANN':   1,\n",
    "        },\n",
    "        fills={\n",
    "            'VQSLOD': np.nan,\n",
    "        },\n",
    "        flatten_filter=True,\n",
    "        progress=100000,\n",
    "        verbose=True,\n",
    "        cache=True,\n",
    "        cachedir=output_dir\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[vcfnp] 2016-07-20 13:33:34.869827 :: caching is enabled\n",
      "[vcfnp] 2016-07-20 13:33:34.870453 :: no cache file found\n",
      "[vcfnp] 2016-07-20 13:33:34.870915 :: building array\n",
      "[vcfnp] 2016-07-20 13:33:39.391394 :: 100000 rows in 4.51s; batch in 4.51s (22181 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:42.435259 :: 200000 rows in 7.55s; batch in 3.04s (32852 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:45.397762 :: 300000 rows in 10.51s; batch in 2.96s (33755 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:48.338666 :: 400000 rows in 13.46s; batch in 2.94s (34003 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:51.069947 :: 500000 rows in 16.19s; batch in 2.73s (36612 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:54.074128 :: 600000 rows in 19.19s; batch in 3.00s (33286 rows/s)\n",
      "[vcfnp] 2016-07-20 13:33:57.401436 :: 700000 rows in 22.52s; batch in 3.33s (30054 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:01.364760 :: 800000 rows in 26.48s; batch in 3.96s (25231 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:05.003762 :: 900000 rows in 30.12s; batch in 3.64s (27480 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:08.831588 :: 1000000 rows in 33.95s; batch in 3.83s (26124 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:11.796042 :: 1100000 rows in 36.91s; batch in 2.96s (33732 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:14.819372 :: 1200000 rows in 39.94s; batch in 3.02s (33076 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:17.760293 :: 1300000 rows in 42.88s; batch in 2.94s (34002 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:20.790095 :: 1400000 rows in 45.91s; batch in 3.03s (33005 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:23.691342 :: 1500000 rows in 48.81s; batch in 2.90s (34467 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:26.808365 :: 1600000 rows in 51.93s; batch in 3.12s (32081 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:29.726791 :: 1700000 rows in 54.84s; batch in 2.92s (34265 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:33.307140 :: 1800000 rows in 58.42s; batch in 3.58s (27930 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:37.268422 :: 1900000 rows in 62.39s; batch in 3.96s (25244 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:40.156891 :: 2000000 rows in 65.27s; batch in 2.89s (34620 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:43.030402 :: 2100000 rows in 68.15s; batch in 2.87s (34800 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:46.006609 :: 2200000 rows in 71.12s; batch in 2.98s (33599 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:48.857667 :: 2300000 rows in 73.97s; batch in 2.85s (35074 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:51.747726 :: 2400000 rows in 76.86s; batch in 2.89s (34601 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:54.864467 :: 2500000 rows in 79.98s; batch in 3.12s (32084 rows/s)\n",
      "[vcfnp] 2016-07-20 13:34:57.611911 :: 2600000 rows in 82.73s; batch in 2.75s (36397 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:00.325527 :: 2700000 rows in 85.44s; batch in 2.71s (36851 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:03.280108 :: 2800000 rows in 88.40s; batch in 2.95s (33845 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:06.238718 :: 2900000 rows in 91.36s; batch in 2.96s (33799 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:09.018021 :: 3000000 rows in 94.14s; batch in 2.78s (35979 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:12.060718 :: 3100000 rows in 97.18s; batch in 3.04s (32865 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:15.221517 :: 3200000 rows in 100.34s; batch in 3.16s (31637 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:18.636966 :: 3300000 rows in 103.75s; batch in 3.42s (29278 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:21.681309 :: 3400000 rows in 106.80s; batch in 3.04s (32847 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:24.984001 :: 3500000 rows in 110.10s; batch in 3.30s (30278 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:28.193292 :: 3600000 rows in 113.31s; batch in 3.21s (31159 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:31.657488 :: 3700000 rows in 116.77s; batch in 3.46s (28866 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:35.121218 :: 3800000 rows in 120.24s; batch in 3.46s (28870 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:38.680401 :: 3900000 rows in 123.80s; batch in 3.56s (28096 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:41.941744 :: 4000000 rows in 127.06s; batch in 3.26s (30662 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:45.735735 :: 4100000 rows in 130.85s; batch in 3.79s (26357 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:49.016738 :: 4200000 rows in 134.13s; batch in 3.28s (30478 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:49.365811 :: 4212465 rows in 134.48s (31323 rows/s)\n",
      "[vcfnp] 2016-07-20 13:35:49.368920 :: saving to cache file /lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz.vcfnp_cache/variants.npy\n"
     ]
    }
   ],
   "source": [
    "create_variants_npy_2(\"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sites_vcf_fn = \"%s/vcf/SNP_INDEL_WG.sites.analysis.vcf.gz\" % (output_dir)\n",
    "variants_all = np.load(\"%s.vcfnp_cache/variants.npy\" % sites_vcf_fn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True b'SNP' True False\n",
      "True b'SNP' True True\n",
      "True b'SNP' False False\n",
      "True b'SNP' False True\n",
      "True b'IND' True False\n",
      "True b'IND' True True\n",
      "True b'IND' False False\n",
      "True b'IND' False True\n",
      "False b'SNP' True False\n",
      "False b'SNP' True True\n",
      "False b'SNP' False False\n",
      "False b'SNP' False True\n",
      "False b'IND' True False\n",
      "False b'IND' True True\n",
      "False b'IND' False False\n",
      "False b'IND' False True\n",
      "[[True, b'SNP', True, False, 641856], [True, b'SNP', True, True, 114294], [True, b'SNP', False, False, 417751], [True, b'SNP', False, True, 155428], [True, b'IND', True, False, 7310], [True, b'IND', True, True, 113173], [True, b'IND', False, False, 67700], [True, b'IND', False, True, 417683], [False, b'SNP', True, False, 369312], [False, b'SNP', True, True, 126806], [False, b'SNP', False, False, 424579], [False, b'SNP', False, True, 311864], [False, b'IND', True, False, 29418], [False, b'IND', True, True, 358754], [False, b'IND', False, False, 78760], [False, b'IND', False, True, 577777]]\n"
     ]
    }
   ],
   "source": [
    "counts_list = list()\n",
    "GQ_thresholds = [30, 99, 0]\n",
    "variant_types = [b'SNP', b'IND']\n",
    "multiallelics = [False, True]\n",
    "codings = [True, False]\n",
    "filter_passes = [True, False]\n",
    "\n",
    "for filter_pass in filter_passes:\n",
    "    for variant_type in variant_types:\n",
    "        for coding in codings:\n",
    "            for multiallelic in multiallelics:\n",
    "                print(filter_pass, variant_type, coding, multiallelic)\n",
    "                ix = (\n",
    "                    (variants_all['VARIANT_TYPE'] == variant_type) &\n",
    "                    (variants_all['MULTIALLELIC'] == multiallelic) &\n",
    "                    (variants_all['CDS'] == coding) &\n",
    "                    (variants_all['FILTER_PASS'] == filter_pass)\n",
    "                )\n",
    "                number_of_variants = np.sum(ix)\n",
    "                counts_list.append(\n",
    "                    [filter_pass, variant_type, coding, multiallelic, number_of_variants]\n",
    "                )\n",
    "\n",
    "print(counts_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>PASS</th>\n",
       "<th>Type</th>\n",
       "<th>Coding</th>\n",
       "<th>Multiallelic</th>\n",
       "<th>Variants</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>641856</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>114294</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>417751</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>155428</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>7310</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>113173</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>67700</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>417683</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>369312</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>126806</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>424579</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'SNP'</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>311864</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>29418</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>True</td>\n",
       "<td>True</td>\n",
       "<td>358754</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>78760</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td>b'IND'</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "<td>577777</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "headers = ['PASS', 'Type', 'Coding', 'Multiallelic', 'Variants']\n",
    "etl.wrap(counts_list).pushheader(headers).displayall()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/scratch109/malaria/rp7/data/pf3k/pilot_5_0/20160719_mendelian_error_duplicate_concordance/variant_counts.xlsx'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "etl.wrap(counts_list).pushheader(headers).convert('Variants', int).toxlsx(counts_table_fn)\n",
    "counts_table_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1935195"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(etl\n",
    "       .fromtsv('/nfs/team112_internal/production/release_build/Pf3K/pilot_5_0/passcounts')\n",
    "       .pushheader(('CHROM', 'count'))\n",
    "       .convertnumbers()\n",
    "       .values('count')\n",
    "       .array()\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
